// dev-env-installer/dist/devUtils.js
"use strict";
var modulesDetector = require("./linkedModuleDetector");
var cp = require('child_process');
function execProcess(command, wrkDir, logEnabled, errLogEnabled, messageBefore, messageAfter, messageError, maxLogLength, onError) {
    if (logEnabled === void 0) { logEnabled = false; }
    if (errLogEnabled === void 0) { errLogEnabled = true; }
    if (messageBefore === void 0) { messageBefore = ''; }
    if (messageAfter === void 0) { messageAfter = ''; }
    if (messageError === void 0) { messageError = ''; }
    if (maxLogLength === void 0) { maxLogLength = -1; }
    if (onError === void 0) { onError = null; }
    console.log("> " + wrkDir + " " + command);
    try {
        if (logEnabled) {
            console.log(messageBefore);
        }
        var logObj = cp.execSync(command, {
            cwd: wrkDir,
            encoding: 'utf8',
            stdio: [0, 1, 2]
        });
        if (logEnabled) {
            console.log(messageAfter);
            if (logObj) {
                var log = logObj.toString();
                if (log.trim().length > 0) {
                    if (maxLogLength < 0) {
                        console.log(log);
                    }
                    else if (maxLogLength > 0) {
                        console.log(log.substring(0, Math.min(maxLogLength, log.length)));
                    }
                }
            }
        }
        return 0;
    }
    catch (err) {
        if (onError) {
            onError(err);
        }
        if (errLogEnabled) {
            console.log(messageError);
            console.log(err.message);
        }
        return err.status;
    }
}
exports.execProcess = execProcess;
function pullAll(rootFolder, workspaceDescriptorFile) {
    var modules = modulesDetector.getModules(rootFolder, workspaceDescriptorFile);
    var reversedModules = modules.reverse();
    reversedModules.forEach(function (module) {
        var folder = module.fsLocation;
        if (folder) {
            if (execProcess("git pull", folder, true) != 0) {
                throw new Error("Failed to pull " + folder);
            }
        }
    });
}
exports.pullAll = pullAll;
function buildAll(rootFolder, workspaceDescriptorFile) {
    var modules = modulesDetector.getModules(rootFolder, workspaceDescriptorFile);
    var reversedModules = modules.reverse();
    reversedModules.forEach(function (module) {
        var folder = module.fsLocation;
        if (folder) {
            var buildCommand = module.buildCommand;
            if (buildCommand) {
                if (execProcess(buildCommand, folder, true) != 0) {
                    throw new Error("Failed to build " + folder);
                }
            }
        }
    });
}
exports.buildAll = buildAll;
function testAll(rootFolder, workspaceDescriptorFile) {
    var modules = modulesDetector.getModules(rootFolder, workspaceDescriptorFile);
    var reversedModules = modules.reverse();
    reversedModules.forEach(function (module) {
        var folder = module.fsLocation;
        if (folder) {
            var testCommand = module.testCommand;
            if (testCommand) {
                if (execProcess(testCommand, folder, true) != 0) {
                    throw new Error("Tests failed in " + folder);
                }
            }
        }
    });
}
exports.testAll = testAll;
