// raml-1-0-parser/dist/schema.js
"use strict";
var su = require("./util/schemaUtil");
var refResolvers = require("./raml1/jsyaml/includeRefResolvers");
var schemaGenImpl = require("./raml1/tools/schemaModelGen");
var contentprovider = require('./util/contentprovider');
function createSchema(c, u) {
    return su.createSchema(c, new contentprovider.ContentProvider(u));
}
exports.createSchema = createSchema;
function getXMLSchema(c) {
    return su.getXMLSchema(c);
}
exports.getXMLSchema = getXMLSchema;
function getJSONSchema(c, u) {
    return su.getJSONSchema(c, new contentprovider.ContentProvider(u));
}
exports.getJSONSchema = getJSONSchema;
function completeReference(includePath, includeReference, content) {
    return refResolvers.completeReference(includePath, includeReference, content);
}
exports.completeReference = completeReference;
function getIncludePath(p) {
    return refResolvers.getIncludePath(p);
}
exports.getIncludePath = getIncludePath;
function getIncludeReference(p) {
    return refResolvers.getIncludeReference(p);
}
exports.getIncludeReference = getIncludeReference;
function createSchemaModelGenerator() {
    return new schemaGenImpl.SchemaToModelGenerator();
}
exports.createSchemaModelGenerator = createSchemaModelGenerator;
function createModelToSchemaGenerator() {
    return new schemaGenImpl.ModelToSchemaGenerator();
}
exports.createModelToSchemaGenerator = createModelToSchemaGenerator;
