// raml-1-0-parser/dist/raml1/highLevelImpl.js
/// <reference path="../../typings/main.d.ts" />
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var defs = require("raml-definition-system");
var hl = require("./highLevelAST");
var _ = require("underscore");
var proxy = require("./ast.core/LowLevelASTProxy");
var def = defs;
var builder = require("./ast.core/builder");
var search = require("./ast.core/search");
var mutators = require("./ast.core/mutators");
var linter = require("./ast.core/linter");
var typeBuilder = require("./ast.core/typeBuilder");
var universes = require("./tools/universe");
var jsyaml = require("./jsyaml/jsyaml2lowLevel");
var textutil = require('../util/textutil');
var services = def;
var yaml = require("yaml-ast-parser");
var factory10 = require("./artifacts/raml10factory");
var factory08 = require("./artifacts/raml08factory");
var universeHelpers = require("./tools/universeHelpers");
var resourceRegistry = require("./jsyaml/resourceRegistry");
var rTypes = defs.rt;
var contentprovider = require('../util/contentprovider');
function qName(x, context) {
    var dr = search.declRoot(context);
    var nm = x.name();
    var o = nm;
    var ind = nm.indexOf("<"); //TODO SCOPE IT
    if (ind != -1) {
        nm = nm.substring(0, ind);
    }
    while (true) {
        var np = x.parent();
        if (!np || np == dr) {
            break;
        }
        else {
            if (np.definition().key() == universes.Universe10.Library && np.parent()) {
                nm = np.name() + "." + nm;
            }
            x = np;
        }
    }
    return nm;
}
exports.qName = qName;
var BasicASTNode = (function () {
    function BasicASTNode(_node, _parent) {
        this._node = _node;
        this._parent = _parent;
        this._implicit = false;
        this.values = {};
        if (_node) {
            _node.setHighLevelParseResult(this);
        }
    }
    BasicASTNode.prototype.getKind = function () {
        return hl.NodeKind.BASIC;
    };
    BasicASTNode.prototype.asAttr = function () {
        return null;
    };
    BasicASTNode.prototype.asElement = function () {
        return null;
    };
    BasicASTNode.prototype.hashkey = function () {
        if (!this._hashkey)
            this._hashkey = this.parent() ? this.parent().hashkey() + "/" + this.name() : this.name();
        return this._hashkey;
    };
    BasicASTNode.prototype.root = function () {
        if (this.parent()) {
            return this.parent().root();
        }
        return this;
    };
    BasicASTNode.prototype.version = function () {
        return "";
    };
    BasicASTNode.prototype.getLowLevelStart = function () {
        if (this.lowLevel().kind() === jsyaml.Kind.SCALAR) {
            return this.lowLevel().start();
        }
        return this.lowLevel().keyStart();
    };
    BasicASTNode.prototype.getLowLevelEnd = function () {
        if (this.lowLevel().kind() === jsyaml.Kind.SCALAR) {
            return this.lowLevel().end();
        }
        return this.lowLevel().keyEnd();
    };
    BasicASTNode.prototype.isSameNode = function (n) {
        if (n) {
            if (n.lowLevel().actual() == this.lowLevel().actual()) {
                return true;
            }
        }
        return false;
    };
    BasicASTNode.prototype.checkContextValue = function (name, value, thisObj) {
        var vl = this.computedValue(name);
        if (vl && vl.indexOf(value) != -1) {
            return true; //FIXME
        }
        return value == vl || value == 'false';
    };
    BasicASTNode.prototype.printDetails = function (indent) {
        return (indent ? indent : "") + "Unkown\n";
    };
    /**
     * Used for test comparison of two trees. Touching this will require AST tests update.
     * @param indent
     * @returns {string}
     */
    BasicASTNode.prototype.testSerialize = function (indent) {
        return (indent ? indent : "") + "Unkown\n";
    };
    BasicASTNode.prototype.errors = function () {
        var errors = [];
        var q = {
            accept: function (c) {
                errors.push(c);
            },
            begin: function () {
            },
            end: function () {
            }
        };
        this.validate(q);
        return errors;
    };
    BasicASTNode.prototype.markCh = function () {
        var n = this.lowLevel();
        n = n._node ? n._node : n;
        if (n['markCh']) {
            return true;
        }
        n['markCh'] = 1;
    };
    BasicASTNode.prototype.unmarkCh = function () {
        var n = this.lowLevel();
        n = n._node ? n._node : n;
        delete n['markCh'];
    };
    BasicASTNode.prototype.validate = function (v) {
        linter.validate(this, v);
    };
    BasicASTNode.prototype.allowRecursive = function () {
        return false;
    };
    BasicASTNode.prototype.setComputed = function (name, v) {
        this.values[name] = v;
    };
    BasicASTNode.prototype.computedValue = function (name) {
        var vl = this.values[name];
        if (!vl && this.parent()) {
            return this.parent().computedValue(name);
        }
        return vl;
    };
    BasicASTNode.prototype.lowLevel = function () {
        return this._node;
    };
    BasicASTNode.prototype.name = function () {
        var c = this.lowLevel().key();
        if (!c) {
            return "";
        }
        return c;
    };
    BasicASTNode.prototype.optional = function () {
        var llNode = this.lowLevel();
        var ownValue = llNode.optional();
        if (llNode.key() != null) {
            return ownValue;
        }
        var p = this.property();
        if (!p || !p.isMultiValue()) {
            return ownValue;
        }
        var llParent = llNode.parent();
        while (llParent && llParent.highLevelNode() == null) {
            if (llParent.kind() == yaml.Kind.MAPPING) {
                return llParent.optional();
            }
            llParent = llParent.parent();
        }
        return ownValue;
    };
    BasicASTNode.prototype.parent = function () {
        return this._parent;
    };
    BasicASTNode.prototype.setParent = function (parent) {
        this._parent = parent;
    };
    BasicASTNode.prototype.isElement = function () {
        return false;
    };
    BasicASTNode.prototype.directChildren = function () {
        return this.children();
    };
    BasicASTNode.prototype.children = function () {
        return [];
    };
    BasicASTNode.prototype.isAttached = function () {
        return this.parent() != null;
    };
    BasicASTNode.prototype.isImplicit = function () {
        return this._implicit;
    };
    BasicASTNode.prototype.isAttr = function () {
        return false;
    };
    BasicASTNode.prototype.isUnknown = function () {
        return true;
    };
    BasicASTNode.prototype.id = function () {
        var _this = this;
        if (this._parent) {
            var parentId = this.parent().id();
            parentId += "." + this.name();
            var sameName = this.parent().directChildren().filter(function (x) { return x.name() == _this.name(); });
            if (sameName.length > 1) {
                var ind = sameName.indexOf(this);
                parentId += "[" + ind + "]";
            }
            return parentId;
        }
        return "";
    };
    BasicASTNode.prototype.localId = function () {
        return this.name();
    };
    BasicASTNode.prototype.fullLocalId = function () {
        var _this = this;
        if (this._parent) {
            var result = "." + this.name();
            var sameName = this.parent().directChildren().filter(function (x) { return x.name() == _this.name(); });
            if (sameName.length > 1) {
                var ind = sameName.indexOf(this);
                result += "[" + ind + "]";
            }
            return result;
        }
        return this.localId();
    };
    BasicASTNode.prototype.property = function () {
        return null;
    };
    return BasicASTNode;
}());
exports.BasicASTNode = BasicASTNode;
var StructuredValue = (function () {
    function StructuredValue(node, _parent, _pr, kv) {
        if (kv === void 0) { kv = null; }
        this.node = node;
        this._parent = _parent;
        this.kv = kv;
        this._pr = _pr;
    }
    StructuredValue.prototype.valueName = function () {
        var res = null;
        if (this.kv) {
            res = this.kv;
        }
        res = this.node.key();
        if (this._pr && this._pr.isAnnotation()) {
            if (res && res.charAt(0) == '(') {
                res = res.substring(1, res.length - 1);
            }
        }
        return res;
    };
    StructuredValue.prototype.children = function () {
        return this.node.children().map(function (x) { return new StructuredValue(x, null, null); });
    };
    StructuredValue.prototype.lowLevel = function () {
        return this.node;
    };
    StructuredValue.prototype.toHighLevel = function (parent) {
        if (!parent && this._parent)
            parent = this._parent;
        var vn = this.valueName();
        var cands = search.referenceTargets(this._pr, parent).filter(function (x) { return qName(x, parent) == vn; });
        if (cands && cands[0]) {
            var tp = cands[0].localType();
            var node = new ASTNodeImpl(this.node, parent, tp, this._pr);
            if (this._pr) {
                this._pr.childRestrictions().forEach(function (y) {
                    node.setComputed(y.name, y.value);
                });
            }
            return node;
        }
        //if (this._pr.range()){
        //    var node=new ASTNodeImpl(parent.lowLevel(),parent,this._pr.range(),this._pr);
        //    if (this._pr){
        //        this._pr.childRestrictions().forEach(y=>{
        //            node.setComputed(y.name,y.value)
        //        })
        //    }
        //    return node;
        //}
        return null;
    };
    StructuredValue.prototype.toHighLevel2 = function (parent) {
        if (!parent && this._parent)
            parent = this._parent;
        var vn = this.valueName();
        var cands = search.referenceTargets(this._pr, parent).filter(function (x) { return qName(x, parent) == vn; });
        if (cands && cands[0]) {
            var tp = cands[0].localType();
            var node = new ASTNodeImpl(this.node, parent, tp, this._pr);
            if (this._pr) {
                this._pr.childRestrictions().forEach(function (y) {
                    node.setComputed(y.name, y.value);
                });
            }
            return node;
        }
        if (this._pr.range()) {
            var node = new ASTNodeImpl(this.node.parent(), parent, this._pr.range(), this._pr);
            if (this._pr) {
                this._pr.childRestrictions().forEach(function (y) {
                    node.setComputed(y.name, y.value);
                });
            }
            return node;
        }
        return null;
    };
    return StructuredValue;
}());
exports.StructuredValue = StructuredValue;
var ASTPropImpl = (function (_super) {
    __extends(ASTPropImpl, _super);
    function ASTPropImpl(node, parent, _def, _prop, fromKey) {
        if (fromKey === void 0) { fromKey = false; }
        _super.call(this, node, parent);
        this._def = _def;
        this._prop = _prop;
        this.fromKey = fromKey;
    }
    ASTPropImpl.prototype.definition = function () {
        return this._def;
    };
    ASTPropImpl.prototype.asAttr = function () {
        return this;
    };
    ASTPropImpl.prototype.isString = function () {
        if (this._def) {
            if (this._def.key() === universes.Universe08.StringType || this._def.key() == universes.Universe10.StringType) {
                return true;
            }
        }
        return false;
    };
    ASTPropImpl.prototype.getKind = function () {
        return hl.NodeKind.ATTRIBUTE;
    };
    ASTPropImpl.prototype.owningWrapper = function () {
        return {
            node: this.parent().wrapperNode(),
            property: this.name()
        };
    };
    ASTPropImpl.prototype.patchType = function (t) {
        this._def = t;
    };
    ASTPropImpl.prototype.findReferenceDeclaration = function () {
        var _this = this;
        var targets = search.referenceTargets(this.property(), this.parent());
        var vl = this.value();
        if (vl instanceof StructuredValue) {
            var st = vl;
            var nm = st.valueName();
        }
        else {
            var nm = "" + vl;
        }
        var t = _.find(targets, function (x) { return qName(x, _this.parent()) == nm; });
        return t;
    };
    ASTPropImpl.prototype.findReferencedValue = function () {
        var c = this.findReferenceDeclaration();
        if (c) {
            var vl = c.attr("value");
            var ck = c.definition().key();
            if (ck === universes.Universe08.GlobalSchema || ck === universes.Universe10.GlobalSchema) {
                if (vl) {
                    var actualValue = vl.value();
                    if (actualValue) {
                        var rf = linter.isValid(this._def, this.parent(), actualValue, vl.property());
                        return rf;
                    }
                }
                return null;
            }
        }
        return c;
    };
    ASTPropImpl.prototype.isElement = function () {
        return false;
    };
    ASTPropImpl.prototype.property = function () {
        return this._prop;
    };
    ASTPropImpl.prototype.convertMultivalueToString = function (value) {
        //|\n  xxx\n  yyy\n  zzz
        var gap = 0;
        var pos = 2;
        while (value[pos] == ' ') {
            gap++;
            pos++;
        }
        //console.log('gap: ' + gap);
        var lines = textutil.splitOnLines(value);
        lines = lines.map(function (line) {
            //console.log('line: ' + line);
            return line.substring(gap, line.length);
        });
        return lines.join('');
    };
    ASTPropImpl.prototype.overrideValue = function (value) {
        this._value = value;
    };
    ASTPropImpl.prototype.value = function () {
        if (this._value) {
            return this._value;
        }
        if (this._computed) {
            return this.computedValue(this.property().nameId());
        }
        if (this.fromKey) {
            return this._node.key();
        }
        if (this.property().isAnnotation() && this._node.key() && this._node.key() != 'annotations') {
            return new StructuredValue(this._node, this.parent(), this._prop);
        }
        var isString = this.property() != null && universeHelpers.isStringTypeType(this.property().range());
        var actualValue = this._node.value(isString); //TODO FIXME
        if (this.property().isSelfNode()) {
            if (!actualValue || actualValue instanceof jsyaml.ASTNode) {
                actualValue = this._node;
                if (actualValue.children().length == 0) {
                    actualValue = null;
                }
            }
        }
        if (actualValue instanceof jsyaml.ASTNode || actualValue instanceof proxy.LowLevelProxyNode) {
            return new StructuredValue(actualValue, this.parent(), this._prop);
        }
        if (typeof (actualValue) == 'string' && textutil.isMultiLineValue(actualValue)) {
            var res = this.convertMultivalueToString(actualValue);
            //console.log('converted: [' + textutil.replaceNewlines(res) + ']');
            return res;
        }
        return actualValue;
    };
    ASTPropImpl.prototype.name = function () {
        return this._prop.nameId();
    };
    ASTPropImpl.prototype.printDetails = function (indent) {
        var className = this.definition().nameId();
        var definitionClassName = this.property().range().nameId();
        var result = (indent ? indent : "") +
            (this.name() + " : " + className
                + "[" + definitionClassName + "]"
                + "  =  " + this.value()) +
            "\n";
        if (this.value() instanceof StructuredValue) {
            var structuredHighLevel = this.value().toHighLevel();
            if (structuredHighLevel && structuredHighLevel.printDetails) {
                result += structuredHighLevel.printDetails(indent + "\t");
            }
        }
        return result;
    };
    /**
     * Used for test comparison of two trees. Touching this will require AST tests update.
     * @param indent
     * @returns {string}
     */
    ASTPropImpl.prototype.testSerialize = function (indent) {
        var className = this.definition().nameId();
        var result = (indent ? indent : "") +
            (this.name() + " : " + className
                + "  =  " + this.value()) +
            "\n";
        if (this.value() instanceof StructuredValue) {
            var structuredHighLevel = this.value().toHighLevel();
            if (structuredHighLevel && structuredHighLevel.testSerialize) {
                result += structuredHighLevel.testSerialize((indent ? indent : "") + "  ");
            }
            else {
                var lowLevel = this.value().lowLevel();
                var dumpObject = lowLevel.dumpToObject();
                var dump = JSON.stringify(dumpObject);
                var indentedDump = "";
                var dumpLines = dump.split("\n");
                dumpLines.forEach(function (dumpLine) { return indentedDump += ((indent ? indent : "") + "  " + dumpLine + "\n"); });
                result += indentedDump + "\n";
            }
        }
        return result;
    };
    ASTPropImpl.prototype.isAttr = function () {
        return true;
    };
    ASTPropImpl.prototype.isUnknown = function () {
        return false;
    };
    ASTPropImpl.prototype.setValue = function (value) {
        mutators.setValue(this, value);
    };
    ASTPropImpl.prototype.setKey = function (value) {
        mutators.setKey(this, value);
    };
    ASTPropImpl.prototype.children = function () {
        return [];
    };
    ASTPropImpl.prototype.addStringValue = function (value) {
        mutators.addStringValue(this, value);
    };
    ASTPropImpl.prototype.addStructuredValue = function (sv) {
        mutators.addStructuredValue(this, sv);
    };
    ASTPropImpl.prototype.addValue = function (value) {
        if (!this.property().isMultiValue())
            throw new Error("setValue(string) only apply to multi-values properties");
        if (typeof value == 'string') {
            this.addStringValue(value);
        }
        else {
            this.addStructuredValue(value);
        }
    };
    ASTPropImpl.prototype.isEmbedded = function () {
        var keyname = this.lowLevel().asMapping().key.value;
        //console.log('propery: ' + this.property().name());
        //console.log('mapping: ' + keyname);
        return this.property().canBeValue() && keyname != this.property().nameId();
    };
    ASTPropImpl.prototype.remove = function () {
        mutators.removeAttr(this);
    };
    ASTPropImpl.prototype.setValues = function (values) {
        mutators.setValues(this, values);
    };
    ASTPropImpl.prototype.isEmpty = function () {
        if (!this.property().isMultiValue())
            throw new Error("isEmpty() only apply to multi-values attributes");
        //console.log('remove: ' + this.name());
        var node = this.parent();
        var llnode = node.lowLevel();
        //node.lowLevel().show('Parent:');
        var attrs = node.attributes(this.name());
        //console.log('attributes: ' + attrs.length);
        if (attrs.length == 0) {
            return true;
        }
        else if (attrs.length == 1) {
            var anode = attrs[0].lowLevel();
            //console.log('attribute : ' + anode.kindName());
            //anode.show("ATTR:");
            if (anode.isMapping()) {
                // that's crazy but it means zero length array indeed )
                return true;
            }
            else {
                return false;
            }
        }
        else {
            return false;
        }
    };
    return ASTPropImpl;
}(BasicASTNode));
exports.ASTPropImpl = ASTPropImpl;
var nodeBuilder = new builder.BasicNodeBuilder();
(function (OverlayMergeMode) {
    OverlayMergeMode[OverlayMergeMode["MERGE"] = 0] = "MERGE";
    OverlayMergeMode[OverlayMergeMode["AGGREGATE"] = 1] = "AGGREGATE";
})(exports.OverlayMergeMode || (exports.OverlayMergeMode = {}));
var OverlayMergeMode = exports.OverlayMergeMode;
var LowLevelWrapperForTypeSystem = (function () {
    function LowLevelWrapperForTypeSystem(_node) {
        this._node = _node;
    }
    LowLevelWrapperForTypeSystem.prototype.contentProvider = function () {
        var root = this._node && this._node.unit() && ((this._node.includePath && this._node.includePath()) ? this._node.unit().resolve(this._node.includePath()) : this._node.unit());
        return new contentprovider.ContentProvider(root);
    };
    LowLevelWrapperForTypeSystem.prototype.key = function () {
        var vl = this._node.key();
        if (this._node.optional()) {
            vl = vl + "?";
        }
        return vl;
    };
    LowLevelWrapperForTypeSystem.prototype.value = function () {
        if (this._node.valueKind() == yaml.Kind.SEQ) {
            return this.children().map(function (x) { return x.value(); });
        }
        if (this._node.valueKind() == yaml.Kind.MAP) {
            var vl = this._node.dumpToObject(true);
            return vl[this.key()];
        }
        if (this._node.kind() == yaml.Kind.MAP) {
            var vl = this._node.dumpToObject(true);
            return vl;
        }
        return this._node.value();
    };
    LowLevelWrapperForTypeSystem.prototype.children = function () {
        return this._node.children().map(function (x) { return new LowLevelWrapperForTypeSystem(x); });
    };
    LowLevelWrapperForTypeSystem.prototype.childWithKey = function (k) {
        var mm = this._node.children();
        for (var i = 0; i < mm.length; i++) {
            if (mm[i].key() == k) {
                return new LowLevelWrapperForTypeSystem(mm[i]);
            }
        }
        return null;
    };
    LowLevelWrapperForTypeSystem.prototype.kind = function () {
        if (this._node.valueKind() == yaml.Kind.MAPPING) {
            return rTypes.NodeKind.MAP;
        }
        if (this._node.valueKind() == yaml.Kind.MAP) {
            return rTypes.NodeKind.MAP;
        }
        if (this._node.kind() == yaml.Kind.MAP) {
            return rTypes.NodeKind.MAP;
        }
        if (this._node.valueKind() == yaml.Kind.SEQ) {
            return rTypes.NodeKind.ARRAY;
        }
        if (this._node.valueKind() == yaml.Kind.INCLUDE_REF) {
            if (this._node.children().length > 0) {
                //we can safely assume that it is map in the type system in this case
                return rTypes.NodeKind.MAP;
            }
        }
        return rTypes.NodeKind.SCALAR;
    };
    return LowLevelWrapperForTypeSystem;
}());
exports.LowLevelWrapperForTypeSystem = LowLevelWrapperForTypeSystem;
var ASTNodeImpl = (function (_super) {
    __extends(ASTNodeImpl, _super);
    function ASTNodeImpl(node, parent, _def, _prop) {
        _super.call(this, node, parent);
        this._def = _def;
        this._prop = _prop;
        this._expanded = false;
        this._allowQuestion = false;
        this._subTypesCache = null;
        this._auxChecked = false;
        /**
         * Depending on the merge mode, overlays and extensions are either merged with the master, or their trees are joined via aggregation
         * @type {OverlayMergeMode}
         */
        this.overlayMergeMode = OverlayMergeMode.MERGE;
        if (node) {
            node.setHighLevelNode(this);
        }
    }
    ASTNodeImpl.prototype.types = function () {
        var _this = this;
        if (!this._types) {
            if (this.parent() && (this.definition().key() !== universes.Universe10.Library)) {
                return this.parent().types();
            }
            else {
                var c = this.lowLevel().actual();
                if (c.types) {
                    return c.types;
                }
                this._types = rTypes.parseFromAST(new LowLevelWrapperForTypeSystem(this.lowLevel()));
                this._types.types().forEach(function (x) {
                    typeBuilder.convertType(_this, x);
                });
                c.types = this._types;
            }
        }
        return this._types;
    };
    ASTNodeImpl.prototype.setTypes = function (t) {
        this._types = t;
    };
    ASTNodeImpl.prototype.parsedType = function () {
        if (!this._ptype) {
            if (this.property() && this.property().nameId() == universes.Universe10.Method.properties.body.name) {
                this._ptype = rTypes.parseTypeFromAST(this.name(), new LowLevelWrapperForTypeSystem(this.lowLevel()), this.types(), true);
            }
            else {
                var annotation = this.definition().isAssignableFrom(universes.Universe10.AnnotationTypeDeclaration.name);
                this._ptype = rTypes.parseTypeFromAST(this.name(), new LowLevelWrapperForTypeSystem(this.lowLevel()), this.types(), false, annotation);
            }
        }
        return this._ptype;
    };
    ASTNodeImpl.prototype.localType = function () {
        return typeBuilder.typeFromNode(this);
    };
    ASTNodeImpl.prototype.patchProp = function (pr) {
        this._prop = pr;
    };
    ASTNodeImpl.prototype.getKind = function () {
        return hl.NodeKind.NODE;
    };
    ASTNodeImpl.prototype.wrapperNode = function () {
        if (!this._wrapperNode) {
            //forcing discrimination
            this.children();
            this._wrapperNode = this.buildWrapperNode();
        }
        return this._wrapperNode;
    };
    ASTNodeImpl.prototype.asElement = function () {
        return this;
    };
    ASTNodeImpl.prototype.buildWrapperNode = function () {
        var ramlVersion = this.definition().universe().version();
        if (ramlVersion == 'RAML10') {
            return factory10.buildWrapperNode(this);
        }
        else if (ramlVersion == 'RAML08') {
            return factory08.buildWrapperNode(this);
        }
        return null;
    };
    ASTNodeImpl.prototype.propertiesAllowedToUse = function () {
        var _this = this;
        return this.definition().allProperties().filter(function (x) { return _this.isAllowedToUse(x); });
    };
    ASTNodeImpl.prototype.isAllowedToUse = function (p) {
        var _this = this;
        var ok = true;
        if (p.getAdapter(services.RAMLPropertyService).isSystem()) {
            return false;
        }
        p.getContextRequirements().forEach(function (y) {
            if (y.name.indexOf('(') != -1) {
                //TODO HANDLE IT LATER
                return true;
            }
            var vl = _this.computedValue(y.name);
            if (vl) {
                ok = ok && (vl == y.value);
            }
            else {
                if (y.value) {
                    ok = false;
                }
            }
        });
        return ok;
    };
    ASTNodeImpl.prototype.allowRecursive = function () {
        if (this.definition().getAdapter(services.RAMLService).isUserDefined()) {
            return true;
        }
        return false;
    };
    ASTNodeImpl.prototype.setWrapperNode = function (node) {
        this._wrapperNode = node;
    };
    ASTNodeImpl.prototype.setAssociatedType = function (d) {
        this._associatedDef = d;
    };
    ASTNodeImpl.prototype.associatedType = function () {
        return this._associatedDef;
    };
    ASTNodeImpl.prototype.knownIds = function () {
        //initializing ids if needed
        //TODO refactor workaround
        this.isAuxilary();
        if (this._knownIds) {
            return this._knownIds;
        }
        else {
            return {};
        }
    };
    ASTNodeImpl.prototype.findById = function (id) {
        var _this = this;
        var v = this._knownIds;
        if (!v) {
            this._knownIds = {};
            var all = search.allChildren(this);
            all.forEach(function (x) { return _this._knownIds[x.id()] = x; });
        }
        return this._knownIds[id];
    };
    ASTNodeImpl.prototype.isAuxilary = function () {
        if (this._isAux) {
            return true;
        }
        if (this._auxChecked) {
            return false;
        }
        this._auxChecked = true;
        var masterApi = this.getMaster();
        if (masterApi) {
            this._isAux = true;
            this.initilizeKnownIDs(masterApi);
            return true;
        }
        return false;
    };
    ASTNodeImpl.prototype.initilizeKnownIDs = function (api) {
        var _this = this;
        this._knownIds = {};
        var allChildren = search.allChildren(api);
        allChildren.forEach(function (x) { return _this._knownIds[x.id()] = x; });
        this._knownIds[""] = api;
    };
    ASTNodeImpl.prototype.getMaster = function () {
        if (this.masterApi) {
            return this.masterApi;
        }
        return this.calculateMasterByRef();
    };
    /**
     * Forcefully sets a master unit for this API, which may be different from the one, current unit points to
     * via masterRef.
     * @param master
     */
    ASTNodeImpl.prototype.overrideMaster = function (master) {
        this.masterApi = master;
        this.resetAuxilaryState();
    };
    ASTNodeImpl.prototype.setMergeMode = function (mergeMode) {
        this.overlayMergeMode = mergeMode;
        this.resetAuxilaryState();
    };
    ASTNodeImpl.prototype.getMergeMode = function () {
        return this.overlayMergeMode;
    };
    ASTNodeImpl.prototype.calculateMasterByRef = function () {
        var unit = this.lowLevel().unit();
        if (!unit)
            return null;
        var masterReferenceNode = unit.getMasterReferenceNode();
        if (!masterReferenceNode || !masterReferenceNode.value()) {
            return null;
        }
        var masterPath = masterReferenceNode.value();
        var masterUnit = this.lowLevel().unit().project().resolve(this.lowLevel().unit().path(), masterPath);
        if (!masterUnit) {
            return null;
        }
        var result = fromUnit(masterUnit);
        result.setMergeMode(this.overlayMergeMode);
        return result;
    };
    ASTNodeImpl.prototype.resetAuxilaryState = function () {
        this._isAux = false;
        this._auxChecked = false;
        this._knownIds = null;
        this.clearChildrenCache();
    };
    ASTNodeImpl.prototype.printDetails = function (indent) {
        var result = "";
        if (!indent)
            indent = "";
        var classname = this.definition().nameId();
        var definitionClasName = this.property() ? this.property().range().nameId() : "";
        var parentPropertyName = this.property() ? this.property().nameId() : "";
        result += indent + classname + "[" + definitionClasName + "]" + " <--- " + parentPropertyName + "\n";
        this.children().forEach(function (child) {
            result += child.printDetails(indent + "\t");
        });
        return result;
    };
    /**
     * Used for test comparison of two trees. Touching this will require AST tests update.
     * @param indent
     * @returns {string}
     */
    ASTNodeImpl.prototype.testSerialize = function (indent) {
        var result = "";
        if (!indent)
            indent = "";
        var classname = this.definition().nameId();
        var parentPropertyName = this.property() ? this.property().nameId() : "";
        result += indent + classname + " <-- " + parentPropertyName + "\n";
        this.children().forEach(function (child) {
            if (child.testSerialize) {
                result += child.testSerialize(indent + "  ");
            }
        });
        return result;
    };
    ASTNodeImpl.prototype.getExtractedChildren = function () {
        var r = this.root();
        if (r.isAuxilary()) {
            if (r._knownIds) {
                var i = r._knownIds[this.id()];
                if (i) {
                    return i.children();
                }
            }
            return [];
        }
        return [];
    };
    ASTNodeImpl.prototype.allowsQuestion = function () {
        return this._allowQuestion || this.definition().getAdapter(services.RAMLService).getAllowQuestion();
    };
    ASTNodeImpl.prototype.findReferences = function () {
        var _this = this;
        var rs = [];
        search.refFinder(this.root(), this, rs);
        //TODO FIX ME
        if (rs.length > 1) {
            rs = rs.filter(function (x) { return x != _this && x.parent() != _this; });
        }
        //filtering out equal results
        var filteredReferences = [];
        rs.forEach(function (ref) {
            if (!_.find(filteredReferences, function (existing) { return existing == ref; })) {
                filteredReferences.push(ref);
            }
        });
        return filteredReferences;
    };
    ASTNodeImpl.prototype.setNamePatch = function (s) {
        this._patchedName = s;
    };
    ASTNodeImpl.prototype.isNamePatch = function () {
        return this._patchedName;
    };
    ASTNodeImpl.prototype.name = function () {
        if (this._patchedName) {
            return this._patchedName;
        }
        var ka = _.find(this.directChildren(), function (x) { return x.property() && x.property().getAdapter(services.RAMLPropertyService).isKey(); });
        if (ka && ka instanceof ASTPropImpl) {
            var c = ka.value();
            if (c) {
                var io = c.indexOf(':');
                if (io != -1) {
                    return c.substring(0, io);
                }
            }
            return c;
        }
        return _super.prototype.name.call(this);
    };
    ASTNodeImpl.prototype.findElementAtOffset = function (n) {
        return this._findNode(this, n, n);
    };
    ASTNodeImpl.prototype.isElement = function () {
        return true;
    };
    ASTNodeImpl.prototype.universe = function () {
        if (this._universe) {
            return this._universe;
        }
        return this.definition().universe();
    };
    ASTNodeImpl.prototype.setUniverse = function (u) {
        this._universe = u;
    };
    ASTNodeImpl.prototype._findNode = function (n, offset, end) {
        var _this = this;
        if (n == null) {
            return null;
        }
        if (n.lowLevel()) {
            //var node:ASTNode=<ASTNode>n;
            if (n.lowLevel().start() <= offset && n.lowLevel().end() >= end) {
                var res = n;
                //TODO INCLUDES
                n.elements().forEach(function (x) {
                    if (x.lowLevel().unit() != n.lowLevel().unit()) {
                        return;
                    }
                    var m = _this._findNode(x, offset, end);
                    if (m) {
                        res = m;
                    }
                });
                return res;
            }
        }
        return null;
    };
    ASTNodeImpl.prototype.isStub = function () {
        return (!this.lowLevel().unit()) || this.lowLevel().unit().isStubUnit();
    };
    ASTNodeImpl.prototype.add = function (node) {
        mutators.addToNode(this, node);
    };
    ASTNodeImpl.prototype.remove = function (node) {
        mutators.removeNodeFrom(this, node);
    };
    ASTNodeImpl.prototype.dump = function (flavor) {
        return this._node.dump();
    };
    ASTNodeImpl.prototype.patchType = function (d) {
        this._def = d;
        var ass = this._associatedDef;
        this._associatedDef = null;
        this._children = null;
    };
    ASTNodeImpl.prototype.children = function () {
        var lowLevel = this.lowLevel();
        if (lowLevel && lowLevel.isValueInclude && lowLevel.isValueInclude() && resourceRegistry.isWaitingFor(lowLevel.includePath())) {
            this._children = null;
            return [];
        }
        if (this._children) {
            return this.mergeChildren(this._children, this.getExtractedChildren());
        }
        if (this._node) {
            this._children = nodeBuilder.process(this, this._node.children());
            this._children = this._children.filter(function (x) { return x != null; });
            //FIXME
            return this.mergeChildren(this._children, this.getExtractedChildren());
        }
        return [];
    };
    ASTNodeImpl.prototype.mergeChildren = function (originalChildren, masterChildren) {
        var _this = this;
        var root = this.root();
        if (root.overlayMergeMode == OverlayMergeMode.AGGREGATE) {
            //simply joining the sets
            return originalChildren.concat(masterChildren);
        }
        else if (root.overlayMergeMode == OverlayMergeMode.MERGE) {
            var result = [];
            originalChildren.forEach(function (originalChild) {
                var masterCounterpart = _.find(masterChildren, function (masterChild) { return masterChild.fullLocalId() == originalChild.fullLocalId(); });
                if (!masterCounterpart) {
                    //we dont have a counterpart, so simply adding to result
                    result.push(originalChild);
                }
                else {
                    //there is a counterpart, so deciding what to do:
                    _this.mergeChild(result, originalChild, masterCounterpart);
                }
            });
            masterChildren.forEach(function (masterChild) {
                var originalCounterpart = _.find(originalChildren, function (originalChild) { return masterChild.fullLocalId() == originalChild.fullLocalId(); });
                if (!originalCounterpart) {
                    //we dont have a counterpart, so simply adding to result
                    result.push(masterChild);
                }
            });
            return result;
        }
        return null;
    };
    ASTNodeImpl.prototype.mergeChild = function (result, originalChild, masterChild) {
        if (originalChild.getKind() != masterChild.getKind()) {
            //should not happen theoretically
            result.push(originalChild);
            result.push(masterChild);
            return;
        }
        if (originalChild.getKind() == hl.NodeKind.NODE) {
            result.push(originalChild);
            return;
        }
        else if (originalChild.getKind() == hl.NodeKind.ATTRIBUTE) {
            //if ((<ASTPropImpl>originalChild).name() == "displayName" ||
            //    (<ASTPropImpl>originalChild).name() == "title") {
            //    console.log("OriginalChildForDisplayName: " + (<ASTPropImpl>originalChild).value())
            //    console.log("MasterChildForDisplayName: " + (<ASTPropImpl>masterChild).value())
            //
            //}
            result.push(originalChild);
            return;
        }
        else if (originalChild.getKind() == hl.NodeKind.BASIC) {
            //we do not know what to do with basic nodes, so adding both.
            result.push(originalChild);
            result.push(masterChild);
            return;
        }
    };
    ASTNodeImpl.prototype.directChildren = function () {
        if (this._children) {
            return this._children;
        }
        if (this._node) {
            this._children = nodeBuilder.process(this, this._node.children());
            return this._children;
        }
        return [];
    };
    ASTNodeImpl.prototype.resetChildren = function () {
        this._children = null;
    };
    ASTNodeImpl.prototype.isEmptyRamlFile = function () {
        var llroot = this.lowLevel().root();
        return llroot.isScalar();
    };
    ASTNodeImpl.prototype.initRamlFile = function () {
        mutators.initEmptyRAMLFile(this);
    };
    ASTNodeImpl.prototype.createAttr = function (n, v) {
        mutators.createAttr(this, n, v);
    };
    ASTNodeImpl.prototype.isAttr = function () {
        return false;
    };
    ASTNodeImpl.prototype.isUnknown = function () {
        return false;
    };
    ASTNodeImpl.prototype.value = function () {
        return this._node.value();
    };
    ASTNodeImpl.prototype.valuesOf = function (propName) {
        var pr = this._def.property(propName);
        if (pr != null) {
            return this.elements().filter(function (x) { return x.property() == pr; });
        }
        return [];
    };
    ASTNodeImpl.prototype.attr = function (n) {
        return _.find(this.attrs(), function (y) { return y.name() == n; });
    };
    ASTNodeImpl.prototype.attrOrCreate = function (name) {
        var a = this.attr(name);
        if (!a)
            this.createAttr(name, '');
        return this.attr(name);
    };
    ASTNodeImpl.prototype.attrValue = function (n) {
        var a = this.attr(n);
        return a ? a.value() : null;
    };
    ASTNodeImpl.prototype.attributes = function (n) {
        return _.filter(this.attrs(), function (y) { return y.name() == n; });
    };
    ASTNodeImpl.prototype.attrs = function () {
        var rs = this.children().filter(function (x) { return x.isAttr(); });
        if (this._patchedName) {
            var kp = _.find(this.definition().allProperties(), function (x) { return x.getAdapter(services.RAMLPropertyService).isKey(); });
            var mm = new ASTPropImpl(this.lowLevel(), this, kp.range(), kp, true);
            mm._value = this._patchedName;
            return [mm].concat(rs);
        }
        return rs;
    };
    ASTNodeImpl.prototype.elements = function () {
        return this.children()
            .filter(function (x) { return !x.isAttr() && !x.isUnknown(); });
    };
    ASTNodeImpl.prototype.element = function (n) {
        var r = this.elementsOfKind(n);
        if (r.length > 0) {
            return r[0];
        }
        return null;
    };
    ASTNodeImpl.prototype.elementsOfKind = function (n) {
        var r = this.elements().filter(function (x) { return x.property().nameId() == n; });
        return r;
    };
    ASTNodeImpl.prototype.definition = function () {
        return this._def;
    };
    ASTNodeImpl.prototype.property = function () {
        return this._prop;
    };
    ASTNodeImpl.prototype.isExpanded = function () {
        return this._expanded;
    };
    ASTNodeImpl.prototype.copy = function () {
        return new ASTNodeImpl(this.lowLevel().copy(), this.parent(), this.definition(), this.property());
    };
    ASTNodeImpl.prototype.clearChildrenCache = function () {
        this._children = null;
    };
    ASTNodeImpl.prototype.optionalProperties = function () {
        var def = this.definition();
        if (def == null) {
            return [];
        }
        var result = [];
        var map = {};
        var children = this.lowLevel().children();
        children.forEach(function (x) {
            if (x.optional()) {
                map[x.key()] = true;
            }
        });
        var props = def.allProperties();
        props.forEach(function (x) {
            var prop = x;
            if (map[prop.nameId()]) {
                result.push(prop.nameId());
            }
        });
        return result;
    };
    return ASTNodeImpl;
}(BasicASTNode));
exports.ASTNodeImpl = ASTNodeImpl;
exports.universeProvider = require("./definition-system/universeProvider");
var getDefinitionSystemType = function (contents, ast) {
    var spec = "";
    var ptype = "Api";
    var num = 0;
    var pt = 0;
    for (var n = 0; n < contents.length; n++) {
        var c = contents.charAt(n);
        if (c == '\r' || c == '\n') {
            if (spec) {
                ptype = contents.substring(pt, n).trim();
            }
            else {
                spec = contents.substring(0, n).trim();
            }
            break;
        }
        if (c == ' ') {
            num++;
            if (!spec && num == 2) {
                spec = contents.substring(0, n);
                pt = n;
            }
        }
    }
    var localUniverse = spec == "#%RAML 1.0" ? new def.Universe(null, "RAML10", exports.universeProvider("RAML10"), "RAML10") : new def.Universe(null, "RAML08", exports.universeProvider("RAML08"));
    if (ptype == 'API') {
        ptype = "Api";
    }
    if (ptype == 'NamedExample') {
        ptype = "ExampleSpec";
    }
    if (ptype == 'DataType') {
        ptype = "TypeDeclaration";
    }
    if (ptype == 'SecurityScheme') {
        ptype = "AbstractSecurityScheme";
    }
    localUniverse.setTopLevel(ptype);
    localUniverse.setTypedVersion(spec);
    // localUniverse.setDescription(spec);
    return { ptype: ptype, localUniverse: localUniverse };
};
function getFragmentDefenitionName(highLevelNode) {
    var contents = highLevelNode.lowLevel() && highLevelNode.lowLevel().unit() && highLevelNode.lowLevel().unit().contents();
    if (contents == null) {
        return null;
    }
    return getDefinitionSystemType(contents, highLevelNode.lowLevel()).ptype;
}
exports.getFragmentDefenitionName = getFragmentDefenitionName;
function fromUnit(l) {
    if (l == null)
        return null;
    var contents = l.contents();
    var ast = l.ast();
    var __ret = getDefinitionSystemType(contents, ast);
    var ptype = __ret.ptype;
    var localUniverse = __ret.localUniverse;
    var apiType = localUniverse.type(ptype);
    if (!apiType)
        apiType = localUniverse.type("Api");
    var api = new ASTNodeImpl(ast, null, apiType, null);
    api.setUniverse(localUniverse);
    //forcing discrimination
    api.children();
    return api;
}
exports.fromUnit = fromUnit;
