// raml-1-0-parser/dist/raml1/lowLevelAST.js
"use strict";
var ASTDelta = (function () {
    function ASTDelta() {
    }
    return ASTDelta;
}());
exports.ASTDelta = ASTDelta;
(function (CommandKind) {
    CommandKind[CommandKind["ADD_CHILD"] = 0] = "ADD_CHILD";
    CommandKind[CommandKind["REMOVE_CHILD"] = 1] = "REMOVE_CHILD";
    CommandKind[CommandKind["MOVE_CHILD"] = 2] = "MOVE_CHILD";
    CommandKind[CommandKind["CHANGE_KEY"] = 3] = "CHANGE_KEY";
    CommandKind[CommandKind["CHANGE_VALUE"] = 4] = "CHANGE_VALUE";
    CommandKind[CommandKind["INIT_RAML_FILE"] = 5] = "INIT_RAML_FILE";
})(exports.CommandKind || (exports.CommandKind = {}));
var CommandKind = exports.CommandKind;
var TextChangeCommand = (function () {
    function TextChangeCommand(offset, replacementLength, text, unit, target) {
        if (target === void 0) { target = null; }
        this.offset = offset;
        this.replacementLength = replacementLength;
        this.text = text;
        this.unit = unit;
        this.target = target;
    }
    return TextChangeCommand;
}());
exports.TextChangeCommand = TextChangeCommand;
var CompositeCommand = (function () {
    function CompositeCommand() {
        this.commands = [];
    }
    return CompositeCommand;
}());
exports.CompositeCommand = CompositeCommand;
(function (InsertionPointType) {
    InsertionPointType[InsertionPointType["NONE"] = 0] = "NONE";
    InsertionPointType[InsertionPointType["START"] = 1] = "START";
    InsertionPointType[InsertionPointType["END"] = 2] = "END";
    InsertionPointType[InsertionPointType["POINT"] = 3] = "POINT";
})(exports.InsertionPointType || (exports.InsertionPointType = {}));
var InsertionPointType = exports.InsertionPointType;
var ASTChangeCommand = (function () {
    function ASTChangeCommand(kind, target, value, position) {
        this.toSeq = false;
        this.kind = kind;
        this.target = target;
        this.value = value;
        this.position = position;
    }
    return ASTChangeCommand;
}());
exports.ASTChangeCommand = ASTChangeCommand;
function setAttr(t, value) {
    return new ASTChangeCommand(CommandKind.CHANGE_VALUE, t, value, -1);
}
exports.setAttr = setAttr;
function setAttrStructured(t, value) {
    return new ASTChangeCommand(CommandKind.CHANGE_VALUE, t, value.lowLevel(), -1);
}
exports.setAttrStructured = setAttrStructured;
function setKey(t, value) {
    return new ASTChangeCommand(CommandKind.CHANGE_KEY, t, value, -1);
}
exports.setKey = setKey;
function removeNode(t, child) {
    return new ASTChangeCommand(CommandKind.REMOVE_CHILD, t, child, -1);
}
exports.removeNode = removeNode;
function insertNode(t, child, insertAfter, toSeq) {
    if (insertAfter === void 0) { insertAfter = null; }
    if (toSeq === void 0) { toSeq = false; }
    var s = new ASTChangeCommand(CommandKind.ADD_CHILD, t, child, -1);
    s.insertionPoint = insertAfter;
    s.toSeq = toSeq;
    return s;
}
exports.insertNode = insertNode;
function initRamlFile(root, newroot) {
    return new ASTChangeCommand(CommandKind.INIT_RAML_FILE, root, newroot, -1);
}
exports.initRamlFile = initRamlFile;
var LineMapperImpl = (function () {
    function LineMapperImpl(content, absPath) {
        this.content = content;
        this.absPath = absPath;
    }
    LineMapperImpl.prototype.position = function (_pos) {
        var pos = _pos;
        this.initMapping();
        for (var i = 0; i < this.mapping.length; i++) {
            var lineLength = this.mapping[i];
            if (pos < lineLength) {
                return {
                    line: i,
                    column: pos,
                    position: _pos
                };
            }
            pos -= lineLength;
        }
        if (pos == 0) {
            return {
                line: this.mapping.length - 1,
                column: this.mapping[this.mapping.length - 1],
                position: this.content.length
            };
        }
        if (pos == 1) {
            //sometimes YAML library reports an error at a position of document length + 1, no idea what they want
            //to tell us that way
            return {
                line: this.mapping.length - 1,
                column: this.mapping[this.mapping.length - 1] - 1,
                position: _pos - 1
            };
        }
        throw new Error("Character position exceeds text length: " + _pos + " > + " + this.content.length + ".\nUnit path: " + this.absPath);
    };
    LineMapperImpl.prototype.initMapping = function () {
        if (this.mapping != null) {
            return;
        }
        if (this.content == null) {
            throw new Error("Line Mapper has been given null content" + (this.absPath != null
                ? ('. Path: ' + this.absPath) : ' and null path.'));
        }
        this.mapping = [];
        var ind = 0;
        var l = this.content.length;
        for (var i = 0; i < l; i++) {
            if (this.content.charAt(i) == '\r') {
                if (i < l - 1 && this.content.charAt(i + 1) == '\n') {
                    this.mapping.push(i - ind + 2);
                    ind = i + 2;
                    i++;
                }
                else {
                    this.mapping.push(i - ind + 1);
                    ind = i + 1;
                }
            }
            else if (this.content.charAt(i) == '\n') {
                this.mapping.push(i - ind + 1);
                ind = i + 1;
            }
        }
        this.mapping.push(l - ind);
    };
    return LineMapperImpl;
}());
exports.LineMapperImpl = LineMapperImpl;
