// raml-typesystem/dist/src/typeExpressions.js
"use strict";
var typeExpression = require("./typeExpressionParser");
var ts = require("./typesystem");
var restrictions_1 = require("./restrictions");
function parseToType(val, t, contentProvider, typeAttributeContentProvider) {
    if (contentProvider === void 0) { contentProvider = null; }
    if (typeAttributeContentProvider === void 0) { typeAttributeContentProvider = null; }
    try {
        var q = val.trim();
        if (q.length > 0) {
            var json = q.charAt(0) == '{';
            if (json || (q.charAt(0) == '<' && q.length > 1 && q.charAt(1) != '<')) {
                return new ts.ExternalType("", q, json, contentProvider, typeAttributeContentProvider);
            }
            var node = typeExpression.parse(val);
            var result = parseNode(node, t);
            return result;
        }
        else {
            return ts.derive(val, [ts.STRING]);
        }
    }
    catch (e) {
        return ts.derive(val, [ts.UNKNOWN]);
    }
}
exports.parseToType = parseToType;
function wrapArray(a, result) {
    while (a > 0) {
        var nt = ts.derive("", [ts.ARRAY]);
        nt.addMeta(new restrictions_1.ComponentShouldBeOfType(result));
        result = nt;
        a--;
    }
    return result;
}
function parseNode(node, t) {
    if (node.type == "union") {
        var ut = node;
        return ts.union("", [parseNode(ut.first, t), parseNode(ut.rest, t)]);
    }
    else if (node.type == "parens") {
        var ps = node;
        var rs = parseNode(ps.expr, t);
        return wrapArray(ps.arr, rs);
    }
    else {
        var lit = node;
        if (lit.value.charAt(lit.value.length - 1) == '?') {
            var result = t.get(lit.value.substr(0, lit.value.length - 1));
            if (!result) {
                result = ts.derive(lit.value, [ts.UNKNOWN]);
            }
            result = ts.union(lit.value, [result, ts.NIL]);
            var a = lit.arr;
            return wrapArray(a, result);
        }
        var result = t.get(lit.value);
        if (!result) {
            result = ts.derive(lit.value, [ts.UNKNOWN]);
        }
        var a = lit.arr;
        return wrapArray(a, result);
    }
}
function storeToString(t) {
    if (t.isSubTypeOf(ts.ARRAY)) {
        var cm = t.oneMeta(restrictions_1.ComponentShouldBeOfType);
        if (cm) {
            if (cm.value().isUnion()) {
                return "(" + storeToString(cm.value()) + ")" + "[]";
            }
            else
                return storeToString(cm.value()) + "[]";
        }
        return "array";
    }
    if (t instanceof ts.UnionType) {
        var ut = t;
        return ut.options().map(function (x) { return storeToString(x); }).join(" | ");
    }
    if (t.isAnonymous()) {
        if (t.isEmpty()) {
            return t.superTypes().map(function (x) { return storeToString(x); }).join(" , ");
        }
    }
    return t.name();
}
exports.storeToString = storeToString;
function visit(node, action) {
    action(node);
    if (node.type == "union") {
        var union = node;
        visit(union.first, action);
        visit(union.rest, action);
    }
    else if (node.type == "parens") {
        var parens = node;
        visit(parens.expr, action);
    }
}
exports.visit = visit;
function serializeToString(node) {
    var arr = 0;
    var str;
    if (node.type == "name") {
        var literal = node;
        str = literal.value;
        arr = literal.arr;
    }
    else if (node.type == "union") {
        var union = node;
        str = serializeToString(union.first) + " | " + serializeToString(union.rest);
    }
    else if (node.type == "parens") {
        var parens = node;
        str = "(" + serializeToString(parens.expr) + ")";
        arr = parens.arr;
    }
    while (--arr >= 0) {
        str += "[]";
    }
    return str;
}
exports.serializeToString = serializeToString;
function parse(str) {
    return typeExpression.parse(str);
}
exports.parse = parse;
