// raml-1-0-parser/dist/raml1/ast.core/expander.js
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var hl = require("../highLevelAST");
var hlimpl = require("../highLevelImpl");
var yaml = require("yaml-ast-parser");
var util = require("../../util/index");
var proxy = require("./LowLevelASTProxy");
var RamlWrapperImpl = require("../artifacts/raml10parser");
var RamlWrapper08Impl = require("../artifacts/raml08parser");
var factory10 = require("../artifacts/raml10factory");
var factory08 = require("../artifacts/raml08factory");
var wrapperHelper = require("../wrapped-ast/wrapperHelper");
var wrapperHelper08 = require("../wrapped-ast/wrapperHelper08");
var pluralize = require("pluralize");
var universeDef = require("../tools/universe");
var _ = require("underscore");
var changeCase = require('change-case');
function expandTraitsAndResourceTypes(api) {
    if (!(api instanceof RamlWrapperImpl.ApiImpl || api instanceof RamlWrapper08Impl.ApiImpl)) {
        return null;
    }
    return new TraitsAndResourceTypesExpander()
        .expandTraitsAndResourceTypes(api);
}
exports.expandTraitsAndResourceTypes = expandTraitsAndResourceTypes;
function mergeAPIs(masterUnit, extensionsAndOverlays, mergeMode) {
    var masterApi = hlimpl.fromUnit(masterUnit);
    if (!masterApi)
        throw new Error("couldn't load api from " + masterUnit.absolutePath());
    if (!extensionsAndOverlays || extensionsAndOverlays.length == 0) {
        return masterApi;
    }
    var highLevelNodes = [];
    for (var i = 0; i < extensionsAndOverlays.length; i++) {
        var unit = extensionsAndOverlays[i];
        var hlNode = hlimpl.fromUnit(unit);
        if (!hlNode) {
            throw new Error("couldn't load api from " + unit.absolutePath());
        }
        highLevelNodes.push(hlNode);
    }
    var lastExtensionOrOverlay = mergeHighLevelNodes(masterApi, highLevelNodes, mergeMode);
    return lastExtensionOrOverlay;
}
exports.mergeAPIs = mergeAPIs;
function mergeHighLevelNodes(masterApi, highLevelNodes, mergeMode) {
    var currentMaster = masterApi;
    highLevelNodes.forEach(function (currentApi) {
        currentApi.overrideMaster(currentMaster);
        currentApi.setMergeMode(mergeMode);
        currentMaster = currentApi;
    });
    return currentMaster;
}
;
var TraitsAndResourceTypesExpander = (function () {
    function TraitsAndResourceTypesExpander() {
    }
    TraitsAndResourceTypesExpander.prototype.expandTraitsAndResourceTypes = function (api) {
        var _this = this;
        if (api.definition().key() == universeDef.Universe10.Overlay) {
            return api;
        }
        this.ramlVersion = api.highLevel().definition().universe().version();
        var factory = this.ramlVersion == "RAML10" ? factory10 : factory08;
        var isRAML1 = api instanceof RamlWrapperImpl.ApiImpl;
        this.globalTraits = this.ramlVersion == "RAML10"
            ? wrapperHelper.allTraits(api)
            : wrapperHelper08.allTraits(api);
        this.globalResourceTypes = this.ramlVersion == "RAML10"
            ? wrapperHelper.allResourceTypes(api)
            : wrapperHelper08.allResourceTypes(api);
        //if ((!traits || traits.length == 0) && (!resourceTypes || resourceTypes.length == 0)) {
        //    return api;
        //}
        if (this.globalTraits.length == 0 && this.globalResourceTypes.length == 0) {
            return api;
        }
        var hlNode = this.createHighLevelNode(api.highLevel());
        var result = factory.buildWrapperNode(hlNode);
        result.setAttributeDefaults(api.getDefaultsCalculator().isEnabled());
        this.traitMap = {};
        this.resourceTypeMap = {};
        result.highLevel().setMergeMode(api.highLevel().getMergeMode());
        var resources = result.resources();
        resources.forEach(function (x) { return _this.processResource(x); });
        return result;
    };
    TraitsAndResourceTypesExpander.prototype.getTemplate = function (name, context, cache, globalList) {
        var unitPath = context.lowLevel().unit().path();
        var unitCache = cache[unitPath];
        if (!unitCache) {
            unitCache = {};
            cache[unitPath] = unitCache;
        }
        var val = unitCache[name];
        if (val !== undefined) {
            return val;
        }
        val = null;
        val = _.find(globalList, function (x) { return hlimpl.qName(x.highLevel(), context) == name; });
        if (!val) {
            val = null;
        }
        unitCache[name] = val;
        return val;
    };
    TraitsAndResourceTypesExpander.prototype.createHighLevelNode = function (api) {
        var highLevelNodes = [];
        var node = api;
        while (node) {
            var llNode = node.lowLevel();
            var topComposite = new proxy.LowLevelCompositeNode(llNode, null, null, this.ramlVersion);
            var nodeType = node.definition();
            var newNode = new hlimpl.ASTNodeImpl(topComposite, null, nodeType, null);
            highLevelNodes.push(newNode);
            node = node.getMaster();
        }
        var masterApi = highLevelNodes.pop();
        highLevelNodes = highLevelNodes.reverse();
        var mergeMode = api.getMergeMode();
        return mergeHighLevelNodes(masterApi, highLevelNodes, mergeMode);
    };
    TraitsAndResourceTypesExpander.prototype.processResource = function (resource) {
        var _this = this;
        var resourceData = this.collectResourceData(resource, resource);
        var resourceLowLevel = resource.highLevel().lowLevel();
        resourceData.filter(function (x) { return x.resourceType != null; }).forEach(function (x) {
            var resourceTypeLowLevel = x.resourceType.node.highLevel().lowLevel();
            var resourceTypeTransformer = new DefaultTransformer(resource, x.resourceType.transformer);
            resourceLowLevel.adopt(resourceTypeLowLevel, resourceTypeTransformer);
        });
        var methods = resource.methods();
        methods.forEach(function (m) {
            var methodLowLevel = m.highLevel().lowLevel();
            var name = m.method();
            resourceData.forEach(function (x) {
                var methodTraits = x.methodTraits[name];
                if (methodTraits) {
                    methodTraits.forEach(function (x) {
                        var traitLowLevel = x.node.highLevel().lowLevel();
                        var traitTransformer = new DefaultTransformer(m, x.transformer);
                        methodLowLevel.adopt(traitLowLevel, traitTransformer);
                    });
                }
                var resourceTraits = x.traits;
                if (resourceTraits) {
                    resourceTraits.forEach(function (x) {
                        var traitLowLevel = x.node.highLevel().lowLevel();
                        var traitTransformer = new DefaultTransformer(m, x.transformer);
                        methodLowLevel.adopt(traitLowLevel, traitTransformer);
                    });
                }
            });
        });
        var resources = resource.resources();
        resources.forEach(function (x) { return _this.processResource(x); });
    };
    TraitsAndResourceTypesExpander.prototype.collectResourceData = function (original, obj, arr, transformer, occuredResourceTypes) {
        var _this = this;
        if (arr === void 0) { arr = []; }
        if (occuredResourceTypes === void 0) { occuredResourceTypes = {}; }
        var ownTraits = this.extractTraits(obj, transformer);
        var methodTraitsMap = {};
        var methods = obj.methods();
        methods.forEach(function (x) {
            var methodTraits = _this.extractTraits(x, transformer);
            if (methodTraits && methodTraits.length > 0) {
                methodTraitsMap[x.method()] = methodTraits;
            }
        });
        var rtData;
        var rt = obj.type();
        if (rt && !occuredResourceTypes[rt.name()]) {
            occuredResourceTypes[rt.name()] = true;
            rtData = this.readGenerictData(original, rt, obj.highLevel(), this.resourceTypeMap, this.globalResourceTypes, 'resource type', transformer);
        }
        arr.push({
            resourceType: rtData,
            traits: ownTraits,
            methodTraits: methodTraitsMap
        });
        if (rtData) {
            this.collectResourceData(original, rtData.node, arr, rtData.transformer, occuredResourceTypes);
        }
        return arr;
    };
    TraitsAndResourceTypesExpander.prototype.extractTraits = function (obj, _transformer, occuredTraits) {
        var _this = this;
        if (occuredTraits === void 0) { occuredTraits = {}; }
        var arr = [];
        for (var i = -1; i < arr.length; i++) {
            var gd = i < 0 ? null : arr[i];
            var _obj = gd ? gd.node : obj;
            var transformer = gd ? gd.transformer : _transformer;
            if (!_obj['is']) {
                continue;
            }
            _obj.is().forEach(function (x) {
                var traitData = _this.readGenerictData(obj, x, _obj.highLevel(), _this.traitMap, _this.globalTraits, 'trait', transformer);
                if (traitData) {
                    var name = traitData.name;
                    //if (!occuredTraits[name]) {
                    occuredTraits[name] = true;
                    arr.push(traitData);
                }
            });
        }
        return arr;
    };
    TraitsAndResourceTypesExpander.prototype.readGenerictData = function (r, obj, context, cache, globalList, template, transformer) {
        var value = obj.value();
        if (typeof (value) == 'string') {
            if (transformer) {
                value = transformer.transform(value).value;
            }
            var node = this.getTemplate(value, context, cache, globalList);
            if (node) {
                return {
                    name: value,
                    transformer: null,
                    node: node
                };
            }
        }
        else if (value instanceof hlimpl.StructuredValue) {
            var sv = value;
            var name = sv.valueName();
            if (transformer) {
                name = transformer.transform(name).value;
            }
            var scalarParams = {};
            var structuredParams = {};
            var node = this.getTemplate(name, context, cache, globalList);
            ;
            //var t = hlimpl.typeFromNode(node.highLevel());
            if (node) {
                if (this.ramlVersion == 'RAML08') {
                    if (transformer) {
                        sv.children().forEach(function (x) { return scalarParams[x.valueName()] = transformer.transform(x.lowLevel().value()).value; });
                    }
                    else {
                        sv.children().forEach(function (x) { return scalarParams[x.valueName()] = x.lowLevel().value(); });
                    }
                }
                else {
                    sv.children().forEach(function (x) {
                        var llNode = x.lowLevel();
                        if (llNode.valueKind() == yaml.Kind.SCALAR) {
                            scalarParams[x.valueName()] = llNode.value();
                        }
                        else {
                            structuredParams[x.valueName()] = llNode;
                        }
                    });
                }
                var ds = new DefaultTransformer(r, null);
                Object.keys(scalarParams).forEach(function (x) {
                    var q = ds.transform(scalarParams[x]);
                    //if (q.value){
                    if (q) {
                        if (typeof q !== "object") {
                            scalarParams[x] = q;
                        }
                    }
                    //}
                });
                return {
                    name: name,
                    transformer: new ValueTransformer(template, name, scalarParams, structuredParams),
                    node: node
                };
            }
        }
        return null;
    };
    return TraitsAndResourceTypesExpander;
}());
var TransformMatches = (function () {
    function TransformMatches(name, transformer) {
        this.name = name;
        this.regexp = new RegExp(TransformMatches.leftTransformRegexp.source + name + TransformMatches.rightTransformRegexp.source);
        this.transformer = transformer;
    }
    TransformMatches.leftTransformRegexp = /\|\s*!\s*/;
    TransformMatches.rightTransformRegexp = /\s*$/;
    return TransformMatches;
}());
var transformers = [
    new TransformMatches("singularize", function (arg) { return pluralize.singular(arg); }),
    new TransformMatches("pluralize", function (arg) { return pluralize.plural(arg); }),
    new TransformMatches("uppercase", function (arg) { return arg ? arg.toUpperCase() : arg; }),
    new TransformMatches("lowercase", function (arg) { return arg ? arg.toLowerCase() : arg; }),
    new TransformMatches("lowercamelcase", function (arg) {
        if (!arg) {
            return arg;
        }
        return changeCase.camelCase(arg);
    }),
    new TransformMatches("uppercamelcase", function (arg) {
        if (!arg) {
            return arg;
        }
        var lowerCamelCase = changeCase.camelCase(arg);
        return lowerCamelCase[0].toUpperCase() + lowerCamelCase.substring(1, lowerCamelCase.length);
    }),
    new TransformMatches("lowerunderscorecase", function (arg) {
        if (!arg) {
            return arg;
        }
        var snakeCase = changeCase.snake(arg);
        return snakeCase.toLowerCase();
    }),
    new TransformMatches("upperunderscorecase", function (arg) {
        if (!arg) {
            return arg;
        }
        var snakeCase = changeCase.snake(arg);
        return snakeCase.toUpperCase();
    }),
    new TransformMatches("lowerhyphencase", function (arg) {
        if (!arg) {
            return arg;
        }
        var paramCase = changeCase.param(arg);
        return paramCase.toLowerCase();
    }),
    new TransformMatches("upperhyphencase", function (arg) {
        if (!arg) {
            return arg;
        }
        var paramCase = changeCase.param(arg);
        return paramCase.toUpperCase();
    })
];
function getTransformNames() {
    return transformers.map(function (transformer) { return transformer.name; });
}
exports.getTransformNames = getTransformNames;
function getTransformerForOccurence(occurence) {
    var result;
    for (var i = 0; i < transformers.length; i++) {
        if (occurence.match(transformers[i].regexp)) {
            result = transformers[i].transformer;
            break;
        }
    }
    return result;
}
exports.getTransformerForOccurence = getTransformerForOccurence;
var TransformationBuffer = (function () {
    function TransformationBuffer() {
        this.buf = null;
    }
    TransformationBuffer.prototype.append = function (value) {
        if (value !== "") {
            if (this.buf != null) {
                if (value != null) {
                    if (typeof (this.buf) != "string") {
                        this.buf = "" + this.buf;
                    }
                    this.buf += value;
                }
            }
            else if (value !== "") {
                this.buf = value;
            }
        }
    };
    TransformationBuffer.prototype.value = function () {
        return this.buf != null ? this.buf : "";
    };
    return TransformationBuffer;
}());
var ValueTransformer = (function () {
    function ValueTransformer(templateKind, templateName, params, structuredParams) {
        this.templateKind = templateKind;
        this.templateName = templateName;
        this.params = params;
        this.structuredParams = structuredParams;
    }
    ValueTransformer.prototype.transform = function (obj, toString) {
        var undefParams = {};
        var errors = [];
        if (typeof (obj) === 'string') {
            if (this.structuredParams && util.stringStartsWith(obj, "<<") && util.stringEndsWith(obj, ">>")) {
                var paramName = obj.substring(2, obj.length - 2);
                var structuredValue = this.structuredParams[paramName];
                if (structuredValue != null) {
                    return { value: structuredValue.value(toString), errors: errors };
                }
            }
            var str = obj;
            var buf = new TransformationBuffer();
            var prev = 0;
            for (var i = str.indexOf('<<'); i >= 0; i = str.indexOf('<<', prev)) {
                buf.append(str.substring(prev, i));
                var i0 = i;
                i += '<<'.length;
                prev = str.indexOf('>>', i);
                if (prev == -1) {
                    break;
                }
                var paramOccurence = str.substring(i, prev);
                prev += '>>'.length;
                var originalString = str.substring(i0, prev);
                var val;
                var paramName;
                var transformer = getTransformerForOccurence(paramOccurence);
                if (transformer) {
                    var ind = paramOccurence.lastIndexOf('|');
                    paramName = paramOccurence.substring(0, ind).trim();
                    val = this.params[paramName];
                    if (val) {
                        val = transformer(val);
                    }
                }
                else {
                    paramName = paramOccurence.trim();
                    val = this.params[paramName];
                }
                if (val === null || val === undefined) {
                    undefParams[paramName] = true;
                    val = originalString;
                }
                buf.append(val);
            }
            var upArr = Object.keys(undefParams);
            if (upArr.length > 0) {
                var errStr = upArr.join(', ').trim();
                var message = "Undefined " + this.templateKind + " parameter" + (upArr.length > 1 ? 's' : '') + ": " + errStr;
                var error = {
                    code: hl.IssueCode.MISSING_REQUIRED_PROPERTY,
                    isWarning: false,
                    message: message,
                    node: null,
                    start: -1,
                    end: -1,
                    path: null
                };
                errors.push(error);
            }
            buf.append(str.substring(prev, str.length));
            return { value: buf.value(), errors: errors };
        }
        else {
            return { value: obj, errors: errors };
        }
    };
    ValueTransformer.prototype.children = function (node) {
        var substitution = this.substitutionNode(node);
        if (substitution) {
            return substitution.children();
        }
        return null;
    };
    ValueTransformer.prototype.valueKind = function (node) {
        var substitution = this.substitutionNode(node);
        if (substitution) {
            return substitution.valueKind();
        }
        return null;
    };
    ValueTransformer.prototype.substitutionNode = function (node) {
        var paramName = this.paramName(node);
        return paramName && this.structuredParams[paramName];
    };
    ValueTransformer.prototype.paramName = function (node) {
        var paramName = null;
        if (node.valueKind() == yaml.Kind.SCALAR) {
            var val = ("" + node.value()).trim();
            if (util.stringStartsWith(val, "<<") && util.stringEndsWith(val, ">>")) {
                paramName = val.substring(2, val.length - 2);
            }
        }
        return paramName;
    };
    return ValueTransformer;
}());
exports.ValueTransformer = ValueTransformer;
var DefaultTransformer = (function (_super) {
    __extends(DefaultTransformer, _super);
    function DefaultTransformer(owner, delegate) {
        _super.call(this, delegate != null ? delegate.templateKind : "", delegate != null ? delegate.templateName : "");
        this.owner = owner;
        this.delegate = delegate;
    }
    DefaultTransformer.prototype.transform = function (obj, toString) {
        if (obj == null) {
            return {
                value: obj,
                errors: []
            };
        }
        var ownResult = {
            value: obj,
            errors: []
        };
        var gotDefaultParam = false;
        defaultParameters.forEach(function (x) { return gotDefaultParam = gotDefaultParam || obj.toString().indexOf('<<' + x) >= 0; });
        if (gotDefaultParam) {
            this.initParams();
            ownResult = _super.prototype.transform.call(this, obj);
        }
        var result = this.delegate != null ? this.delegate.transform(ownResult.value, toString) : ownResult.value;
        return result;
    };
    DefaultTransformer.prototype.initParams = function () {
        var methodName;
        var resourcePath = "";
        var resourcePathName;
        var ll = this.owner.highLevel().lowLevel();
        var node = ll instanceof proxy.LowLevelProxyNode ? ll.originalNode().originalNode() : ll;
        var last = null;
        while (node) {
            var key = node.key();
            if (key != null) {
                if (util.stringStartsWith(key, '/')) {
                    if (!resourcePathName) {
                        var arr = key.split('/');
                        for (var i = arr.length - 1; i >= 0; i--) {
                            var seg = arr[i];
                            if (seg.indexOf('{') == -1) {
                                resourcePathName = arr[i];
                                break;
                            }
                            if (seg.length > 0) {
                                last = seg;
                            }
                        }
                    }
                    resourcePath = key + resourcePath;
                }
                else {
                    methodName = key;
                }
            }
            node = node.parent();
        }
        if (!resourcePathName) {
            if (last) {
                resourcePathName = "";
            }
        }
        this.params = {
            resourcePath: resourcePath,
            resourcePathName: resourcePathName
        };
        if (methodName) {
            this.params['methodName'] = methodName;
        }
    };
    DefaultTransformer.prototype.children = function (node) {
        return this.delegate != null ? this.delegate.children(node) : null;
    };
    DefaultTransformer.prototype.valueKind = function (node) {
        return this.delegate != null ? this.delegate.valueKind(node) : null;
    };
    return DefaultTransformer;
}(ValueTransformer));
exports.DefaultTransformer = DefaultTransformer;
var defaultParameters = ['resourcePath', 'resourcePathName', 'methodName'];
