// raml-1-0-parser/dist/raml1/ast.core/LowLevelASTProxy.js
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var yaml = require("yaml-ast-parser");
var json = require("../jsyaml/json2lowLevel");
var stringify = require("json-stable-stringify");
var impl = require("../jsyaml/jsyaml2lowLevel");
var universes = require("../tools/universe");
var _ = require("underscore");
var LowLevelProxyNode = (function () {
    function LowLevelProxyNode(_parent, _transformer, ramlVersion) {
        this._parent = _parent;
        this._transformer = _transformer;
        this.ramlVersion = ramlVersion;
    }
    LowLevelProxyNode.prototype.keyKind = function () {
        return this._originalNode.keyKind();
    };
    LowLevelProxyNode.prototype.actual = function () {
        if (this._originalNode) {
            return this._originalNode.actual();
        }
        return this;
    };
    LowLevelProxyNode.prototype.transformer = function () { return this._transformer; };
    LowLevelProxyNode.prototype.originalNode = function () {
        return this._originalNode;
    };
    LowLevelProxyNode.prototype.start = function () { return this._originalNode.start(); };
    LowLevelProxyNode.prototype.end = function () { return this._originalNode.end(); };
    LowLevelProxyNode.prototype.value = function (toString) {
        throw new Error('The method must be overridden');
    };
    LowLevelProxyNode.prototype.includeErrors = function () { return this._originalNode.includeErrors(); };
    LowLevelProxyNode.prototype.includePath = function () { return this._originalNode.includePath(); };
    LowLevelProxyNode.prototype.includeReference = function () { return this._originalNode.includeReference(); };
    LowLevelProxyNode.prototype.setKeyOverride = function (_key) {
        this._keyOverride = _key;
    };
    LowLevelProxyNode.prototype.key = function () {
        if (this._keyOverride) {
            return this._keyOverride;
        }
        return this._originalNode.key();
    };
    LowLevelProxyNode.prototype.optional = function () {
        return this.originalNode().optional();
    };
    LowLevelProxyNode.prototype.children = function () {
        throw new Error('The method must be overridden');
    };
    LowLevelProxyNode.prototype.parent = function () { return this._parent; };
    LowLevelProxyNode.prototype.unit = function () { return this._originalNode.unit(); };
    LowLevelProxyNode.prototype.anchorId = function () { return this._originalNode.anchorId(); };
    LowLevelProxyNode.prototype.errors = function () { return this._originalNode.errors(); };
    LowLevelProxyNode.prototype.anchoredFrom = function () { return this._originalNode.anchoredFrom(); };
    LowLevelProxyNode.prototype.includedFrom = function () { return this._originalNode.includedFrom(); };
    LowLevelProxyNode.prototype.visit = function (v) {
        if (v(this)) {
            this.children().forEach(function (x) { return x.visit(v); });
        }
    };
    //TODO
    LowLevelProxyNode.prototype.addChild = function (n) { };
    //TODO
    LowLevelProxyNode.prototype.execute = function (cmd) { };
    //TODO
    LowLevelProxyNode.prototype.dump = function () { return null; };
    LowLevelProxyNode.prototype.dumpToObject = function () {
        var serialized = json.serialize(this);
        if (this.kind() == yaml.Kind.MAPPING) {
            var obj = {};
            obj[this.key()] = serialized;
            return obj;
        }
        return serialized;
    };
    LowLevelProxyNode.prototype.keyStart = function () { return this._originalNode.keyStart(); };
    LowLevelProxyNode.prototype.keyEnd = function () { return this._originalNode.keyEnd(); };
    LowLevelProxyNode.prototype.valueStart = function () { return this._originalNode.valueStart(); };
    LowLevelProxyNode.prototype.valueEnd = function () { return this._originalNode.valueEnd(); };
    LowLevelProxyNode.prototype.isValueLocal = function () { return this._originalNode.isValueLocal(); };
    LowLevelProxyNode.prototype.kind = function () { return this._originalNode.kind(); };
    LowLevelProxyNode.prototype.valueKind = function () { return this._originalNode.valueKind(); };
    LowLevelProxyNode.prototype.show = function (msg) { this._originalNode.show(msg); };
    LowLevelProxyNode.prototype.setHighLevelParseResult = function (highLevelParseResult) {
        this._highLevelParseResult = highLevelParseResult;
    };
    LowLevelProxyNode.prototype.highLevelParseResult = function () {
        return this._highLevelParseResult;
    };
    LowLevelProxyNode.prototype.setHighLevelNode = function (highLevel) {
        this._highLevelNode = highLevel;
    };
    LowLevelProxyNode.prototype.highLevelNode = function () {
        return this._highLevelNode;
    };
    LowLevelProxyNode.prototype.text = function (unitText) {
        throw new Error("not implemented");
    };
    LowLevelProxyNode.prototype.copy = function () {
        throw new Error("not implemented");
    };
    LowLevelProxyNode.prototype.markup = function (json) {
        throw new Error("not implemented");
    };
    LowLevelProxyNode.prototype.nodeDefinition = function () {
        return impl.getDefinitionForLowLevelNode(this);
    };
    LowLevelProxyNode.prototype.includesContents = function () {
        return this._originalNode.includesContents();
    };
    return LowLevelProxyNode;
}());
exports.LowLevelProxyNode = LowLevelProxyNode;
var LowLevelCompositeNode = (function (_super) {
    __extends(LowLevelCompositeNode, _super);
    function LowLevelCompositeNode(node, parent, transformer, ramlVersion, isPrimary) {
        if (isPrimary === void 0) { isPrimary = true; }
        _super.call(this, parent, transformer, ramlVersion);
        this.isPrimary = isPrimary;
        //Colliding nodes of the initioal AST
        this._adoptedNodes = [];
        var originalParent = this.parent() ? this.parent().originalNode() : null;
        this._originalNode = new LowLevelValueTransformingNode(node, originalParent, transformer, this.ramlVersion);
        this._adoptedNodes.push(this._originalNode);
    }
    LowLevelCompositeNode.prototype.adoptedNodes = function () {
        return this._adoptedNodes;
    };
    LowLevelCompositeNode.prototype.primaryNode = function () {
        return this.isPrimary ? this._originalNode : null;
    };
    LowLevelCompositeNode.prototype.parent = function () { return this._parent; };
    LowLevelCompositeNode.prototype.adopt = function (node, transformer) {
        if (!transformer) {
            transformer = this._transformer;
        }
        var originalParent = this.parent() ? this.parent().originalNode() : null;
        var tNode = new LowLevelValueTransformingNode(node, originalParent, transformer, this.ramlVersion);
        this._adoptedNodes.push(tNode);
        if (this._children) {
            this._children.forEach(function (x) { return x._parent = null; });
        }
        this._children = null;
        if (this.highLevelNode()) {
            this.highLevelNode().resetChildren();
        }
    };
    LowLevelCompositeNode.prototype.value = function (toString) {
        var valuableNodes = this._adoptedNodes.filter(function (x) { return x.value() != null; });
        if (valuableNodes.length > 0) {
            return valuableNodes[0].value(toString);
        }
        return this._originalNode.value(toString);
    };
    LowLevelCompositeNode.prototype.children = function () {
        var _this = this;
        if (this._children) {
            return this._children;
        }
        var result = [];
        var canBeMap = false;
        var canBeSeq = false;
        this._adoptedNodes.forEach(function (x) {
            if (x.children() && x.children().length > 0) {
                canBeSeq = true;
                if (x.children()[0].key()) {
                    canBeMap = true;
                }
            }
        });
        if (canBeMap) {
            result = this.collectChildrenWithKeys();
        }
        else if (canBeSeq) {
            result = this.collectChildrenWithKeys();
            var map = {};
            this._adoptedNodes.forEach(function (x) { return x.children().filter(function (y) { return !y.key(); }).forEach(function (y) {
                var isPrimary = x == _this.primaryNode();
                var key = _this.buildKey(y);
                if (!isPrimary && map[key]) {
                    //filtering away values with repeating keys
                    //primary node is not subjected to filtration
                    return;
                }
                map[key] = true;
                var transformer = x.transformer() ? x.transformer() : _this.transformer();
                var ch = (y instanceof LowLevelValueTransformingNode)
                    ? y.originalNode() : y;
                result.push(new LowLevelCompositeNode(ch, _this, transformer, _this.ramlVersion, isPrimary));
            }); });
        }
        else {
            result = [];
        }
        this._children = result;
        return result;
    };
    LowLevelCompositeNode.prototype.buildKey = function (y) {
        var obj = json.serialize(y);
        var def = this.nodeDefinition();
        if (def && (def.key() == universes.Universe08.TraitRef || def.key() == universes.Universe08.ResourceTypeRef
            || def.key() == universes.Universe10.TraitRef || def.key() == universes.Universe10.ResourceTypeRef)) {
            if (typeof obj == 'object') {
                var keys = Object.keys(obj);
                if (keys.length > 0) {
                    obj = keys[0];
                }
            }
        }
        return stringify(obj);
    };
    LowLevelCompositeNode.prototype.collectChildrenWithKeys = function () {
        var _this = this;
        var result = [];
        var m = {};
        this._adoptedNodes.forEach(function (x) {
            var isPrimary = x == _this.primaryNode();
            x.originalNode().children().forEach(function (y) {
                var key = y.key();
                if (key && x.transformer()) {
                    key = x.transformer().transform(key).value;
                }
                if (_this.skipKey(key, isPrimary)) {
                    return;
                }
                if (!key) {
                    return;
                }
                var arr = m[key];
                if (!arr) {
                    arr = [];
                    m[key] = arr;
                }
                arr.push({ node: y, transformer: x.transformer(), isPrimary: isPrimary });
            });
        });
        Object.keys(m).forEach(function (key) {
            var arr = m[key];
            var allOptional = true;
            var hasPrimaryChildren = false;
            arr.forEach(function (x) {
                allOptional = allOptional && x.node.optional();
                hasPrimaryChildren = hasPrimaryChildren || x.isPrimary;
            });
            if (hasPrimaryChildren) {
                var primaryChildren = [];
                arr.filter(function (x) { return x.isPrimary; }).forEach(function (x) {
                    var tr = x.transformer ? x.transformer : _this.transformer();
                    primaryChildren.push(new LowLevelCompositeNode(x.node, _this, tr, _this.ramlVersion, true));
                });
                var primaryChild = primaryChildren[0];
                arr.filter(function (x) { return !x.isPrimary; }).forEach(function (x) {
                    primaryChild.adopt(x.node, x.transformer);
                });
                primaryChildren.forEach(function (x) { return result.push(x); });
            }
            else if (!allOptional) {
                var tr = arr[0].transformer ? arr[0].transformer : _this.transformer();
                var primaryChild = new LowLevelCompositeNode(arr[0].node, _this, tr, _this.ramlVersion, false);
                for (var i = 1; i < arr.length; i++) {
                    primaryChild.adopt(arr[i].node, arr[i].transformer);
                }
                result.push(primaryChild);
            }
        });
        return result;
    };
    LowLevelCompositeNode.prototype.skipKey = function (key, isPrimary) {
        if (isPrimary) {
            return false;
        }
        if (this.ramlVersion != 'RAML08') {
            return false;
        }
        var methodDef = universes.Universe08.Method;
        var hasNormalParametersDef = universes.Universe08.HasNormalParameters;
        var resourceDef = universes.Universe08.Resource;
        //if(key==hasNormalParametersDef.properties.displayName.name
        //    &&this.highLevelNode().definition().key().name==methodDef.name){
        //    return true;
        //}
        //if(key==resourceDef.properties.displayName.name
        //    &&this.highLevelNode().definition().key().name==resourceDef.name){
        //    return true;
        //}
        return false;
    };
    LowLevelCompositeNode.prototype.valueKind = function () {
        if (this._originalNode.kind() != yaml.Kind.MAPPING) {
            return null;
        }
        for (var i = 0; i < this._adoptedNodes.length; i++) {
            var node = this._adoptedNodes[i];
            var yamlNode = node.originalNode()._node;
            if (yamlNode && yamlNode.value != null) {
                return node.valueKind();
            }
        }
        return null;
    };
    LowLevelCompositeNode.prototype.includePath = function () {
        for (var i = 0; i < this._adoptedNodes.length; i++) {
            var node = this._adoptedNodes[i];
            var includePath = node.includePath();
            if (includePath != null) {
                return includePath;
            }
        }
        return null;
    };
    LowLevelCompositeNode.prototype.includeReference = function () {
        for (var i = 0; i < this._adoptedNodes.length; i++) {
            var node = this._adoptedNodes[i];
            if (node.value() != null) {
                return node.includeReference();
            }
        }
        return null;
    };
    LowLevelCompositeNode.prototype.optional = function () {
        return _.all(this._adoptedNodes, function (x) { return x.optional(); });
    };
    return LowLevelCompositeNode;
}(LowLevelProxyNode));
exports.LowLevelCompositeNode = LowLevelCompositeNode;
var LowLevelValueTransformingNode = (function (_super) {
    __extends(LowLevelValueTransformingNode, _super);
    function LowLevelValueTransformingNode(node, parent, transformer, ramlVersion) {
        _super.call(this, parent, transformer, ramlVersion);
        this._originalNode = node;
    }
    LowLevelValueTransformingNode.prototype.value = function (toString) {
        var val = this.originalNode().value(toString);
        var t = this.transformer();
        if (t) {
            var transformationResult = t.transform(val);
            val = transformationResult.value;
        }
        return val;
    };
    LowLevelValueTransformingNode.prototype.children = function () {
        var _this = this;
        return this.originalNode().children().map(function (x) { return new LowLevelValueTransformingNode(x, _this, _this._transformer, _this.ramlVersion); });
    };
    LowLevelValueTransformingNode.prototype.parent = function () { return this._parent; };
    LowLevelValueTransformingNode.prototype.key = function () {
        var key = _super.prototype.key.call(this);
        if (this.transformer() != null) {
            return this.transformer().transform(key).value;
        }
        return key;
    };
    return LowLevelValueTransformingNode;
}(LowLevelProxyNode));
exports.LowLevelValueTransformingNode = LowLevelValueTransformingNode;
