// raml-1-0-parser/dist/raml1/ast.core/namespaceResolver.js
/// <reference path="../../../typings/main.d.ts" />
"use strict";
var ll = require("../lowLevelAST");
var path = require("path");
var universes = require("../tools/universe");
var yaml = require("yaml-ast-parser");
var resourceRegistry = require('../jsyaml/resourceRegistry');
var hlImpl = require("../highLevelImpl");
var def = require("raml-definition-system");
var _ = require("underscore");
var NamespaceResolver = (function () {
    function NamespaceResolver() {
        this.expandedAbsToNsMap = {};
        this._expandedNSMap = {};
        this.byPathMap = {};
        this.byNsMap = {};
        this._hasFragments = {};
        this._unitModels = {};
    }
    NamespaceResolver.prototype.hasTemplates = function (unit) {
        var uModel = this.unitModel(unit);
        if (!uModel.traits.isEmpty() || !uModel.resourceTypes.isEmpty()) {
            return true;
        }
        var epMap = this.expandedPathMap(unit);
        if (epMap) {
            for (var _i = 0, _a = Object.keys(epMap); _i < _a.length; _i++) {
                var p = _a[_i];
                var u = epMap[p].unit;
                var uModel1 = this.unitModel(u);
                if (!uModel1.traits.isEmpty() || !uModel1.resourceTypes.isEmpty()) {
                    return true;
                }
            }
        }
        var eValue = extendsValue(unit);
        while (eValue) {
            unit = unit.resolve(eValue);
            if (unit) {
                uModel = this.unitModel(unit);
                if (!uModel.traits.isEmpty() || !uModel.resourceTypes.isEmpty()) {
                    return true;
                }
                eValue = extendsValue(unit);
            }
            else {
                eValue = null;
            }
        }
        return false;
    };
    NamespaceResolver.prototype.resolveNamespace = function (from, to) {
        if (to == null) {
            return null;
        }
        var toPath = to.absolutePath();
        var unitMap = this.expandedPathMap(from);
        if (!unitMap) {
            return null;
        }
        var usesInfo = unitMap[toPath];
        return usesInfo;
    };
    NamespaceResolver.prototype.expandedNSMap = function (unit) {
        var aPath = unit.absolutePath();
        var result = this._expandedNSMap[aPath];
        if (result === undefined) {
            var pm = this.expandedPathMap(unit);
            if (!pm) {
                result = null;
            }
            else {
                result = {};
                for (var _i = 0, _a = Object.keys(pm); _i < _a.length; _i++) {
                    var ap = _a[_i];
                    var uInfo = pm[ap];
                    result[uInfo.namespace()] = uInfo;
                }
            }
            this._expandedNSMap[aPath] = result;
        }
        return result;
    };
    NamespaceResolver.prototype.expandedPathMap = function (unit) {
        var fromPath = unit.absolutePath();
        var unitMap = this.expandedAbsToNsMap[fromPath];
        if (unitMap === undefined) {
            unitMap = this.calculateExpandedNamespaces(unit);
            if (Object.keys(unitMap).length == 0) {
                unitMap = null;
            }
            this.expandedAbsToNsMap[fromPath] = unitMap;
        }
        return unitMap;
    };
    NamespaceResolver.prototype.calculateExpandedNamespaces = function (_unit) {
        var _this = this;
        var rootPath = path.dirname(_unit.absolutePath());
        var result = {};
        var usesInfoArray = [];
        while (_unit) {
            usesInfoArray.push(new UsesInfo([], _unit, ""));
            var u = _unit;
            _unit = null;
            var node = u.ast();
            if (node && node.kind() != yaml.Kind.SCALAR) {
                var fLine = hlImpl.ramlFirstLine(u.contents());
                if (fLine && fLine.length == 3 && (fLine[2] == def.universesInfo.Universe10.Overlay.name ||
                    fLine[2] == def.universesInfo.Universe10.Extension.name)) {
                    var eValue = extendsValue(u);
                    if (eValue) {
                        _unit = u.resolve(eValue);
                        if (_unit && resourceRegistry.isWaitingFor(_unit.absolutePath())) {
                            _unit = null;
                        }
                    }
                }
            }
        }
        for (var i = 0; i < usesInfoArray.length; i++) {
            var visited = {};
            var usedUnits = {};
            var info = usesInfoArray[i];
            var unit = info.unit;
            var node = unit.ast();
            if (!node || node.kind() == yaml.Kind.SCALAR) {
                continue;
            }
            var steps = info.steps() + 1;
            var visit = function (x) {
                if (!x) {
                    return;
                }
                var children = x.children();
                if (x.parent() == null) {
                    var nodeUnit = x.unit();
                    var localPath = nodeUnit.absolutePath();
                    if (visited[localPath]) {
                        return;
                    }
                    visited[localPath] = true;
                    if (localPath != unit.absolutePath()) {
                        _this._hasFragments[unit.absolutePath()] = true;
                    }
                    var map = _this.pathMap(nodeUnit);
                    if (map) {
                        for (var _i = 0, _a = Object.keys(map); _i < _a.length; _i++) {
                            var absPath = _a[_i];
                            var childInfo = map[absPath];
                            var segments = info.namespaceSegments.concat(childInfo.namespaceSegments);
                            var usesNodes = info.usesNodes.concat(childInfo.usesNodes);
                            var existing = result[absPath];
                            if (existing) {
                                if (existing.steps() < steps) {
                                    continue;
                                }
                                else if (existing.steps() == steps
                                    && _this.lexLessEq(existing.namespaceSegments, segments)) {
                                    continue;
                                }
                            }
                            var includePath;
                            var childInclude = childInfo.includePath;
                            var absChildPath = childInfo.absolutePath();
                            if (path.isAbsolute(info.includePath) || ll.isWebPath(info.includePath)) {
                                includePath = absChildPath;
                            }
                            else if (path.isAbsolute(childInclude) || ll.isWebPath(childInclude)) {
                                includePath = absChildPath;
                            }
                            else {
                                if (ll.isWebPath(rootPath) != ll.isWebPath(absChildPath)) {
                                    includePath = absChildPath;
                                }
                                else if (rootPath.length > 0 && absChildPath.length > 0
                                    && rootPath.charAt(0).toLowerCase() != absChildPath.charAt(0).toLowerCase()) {
                                    //Windows case of library being located on different drive
                                    includePath = absChildPath;
                                }
                                else {
                                    includePath = path.relative(rootPath, absChildPath);
                                }
                            }
                            includePath = includePath.replace(/\\/g, "/");
                            var ui = new UsesInfo(usesNodes, childInfo.unit, includePath);
                            if (!usedUnits[ui.absolutePath()]) {
                                result[absPath] = ui;
                                usesInfoArray.push(ui);
                                usedUnits[ui.absolutePath()] = true;
                            }
                        }
                    }
                }
                children.forEach(function (y) {
                    if (y.includedFrom()) {
                        y = y.parent();
                    }
                    visit(y);
                });
                if (x.parent() == null) {
                    visited[x.unit().absolutePath()] = false;
                }
            };
            visit(unit.ast());
        }
        var namespaces = {};
        for (var _i = 0, _a = Object.keys(result); _i < _a.length; _i++) {
            var key = _a[_i];
            var info = result[key];
            var ns = info.namespace();
            var i = 0;
            while (namespaces[ns]) {
                ns = info.namespace() + i++;
            }
            if (ns != info.namespace()) {
                info.namespaceSegments = ns.split(".");
            }
            namespaces[ns] = true;
        }
        return result;
    };
    NamespaceResolver.prototype.pathMap = function (unit) {
        var fromPath = unit.absolutePath();
        var unitMap = this.byPathMap[fromPath];
        if (unitMap === undefined) {
            unitMap = this.calculateNamespaces(unit);
            if (Object.keys(unitMap).length == 0) {
                unitMap = null;
            }
            this.byPathMap[fromPath] = unitMap;
        }
        return unitMap;
    };
    NamespaceResolver.prototype.nsMap = function (unit) {
        var fromPath = unit.absolutePath();
        var unitMap = this.byNsMap[fromPath];
        if (unitMap === undefined) {
            var map = this.pathMap(unit);
            if (map == null) {
                unitMap = null;
            }
            else {
                unitMap = {};
                for (var _i = 0, _a = Object.keys(map); _i < _a.length; _i++) {
                    var aPath = _a[_i];
                    var info = map[aPath];
                    unitMap[info.namespace()] = info;
                }
            }
            this.byNsMap[fromPath] = unitMap;
        }
        return unitMap;
    };
    NamespaceResolver.prototype.calculateNamespaces = function (unit) {
        var rootPath = path.dirname(unit.absolutePath());
        var result = {};
        var rootNode = unit.ast();
        var usesNodes = rootNode.children().filter(function (x) { return x.key()
            == universes.Universe10.FragmentDeclaration.properties.uses.name; });
        if (rootNode.actual() && rootNode.actual()["usesNode"]) {
            usesNodes = [rootNode.actual()["usesNode"]];
        }
        if (usesNodes.length == 0) {
            return result;
        }
        var usesDeclarationNodes = [];
        for (var _i = 0, usesNodes_1 = usesNodes; _i < usesNodes_1.length; _i++) {
            var un = usesNodes_1[_i];
            usesDeclarationNodes = usesDeclarationNodes.concat(un.children());
        }
        if (usesDeclarationNodes.length == 0) {
            return result;
        }
        for (var _a = 0, usesDeclarationNodes_1 = usesDeclarationNodes; _a < usesDeclarationNodes_1.length; _a++) {
            var un = usesDeclarationNodes_1[_a];
            var value = un.value();
            var libUnit = unit.resolve(value);
            if (libUnit == null) {
                continue;
            }
            var usesNodes = [un];
            var absPath = libUnit.absolutePath();
            var includePath;
            if (path.isAbsolute(value) || ll.isWebPath(value)) {
                includePath = libUnit.absolutePath();
            }
            else {
                includePath = path.relative(rootPath, libUnit.absolutePath());
            }
            includePath = includePath.replace(/\\/g, "/");
            var ui = new UsesInfo(usesNodes, libUnit, includePath);
            result[absPath] = ui;
        }
        var node = unit.ast();
        if (node && node.kind() != yaml.Kind.SCALAR) {
            var fLine = hlImpl.ramlFirstLine(unit.contents());
            if (fLine.length == 3 && (fLine[2] == def.universesInfo.Universe10.Overlay.name ||
                fLine[2] == def.universesInfo.Universe10.Extension.name)) {
                var eNode = _.find(node.children(), function (x) { return x.key() == universes.Universe10.Extension.properties.extends.name; });
                if (eNode) {
                    var eValue = eNode.value(true);
                    var extendedUnit;
                    try {
                        extendedUnit = unit.resolve(eValue);
                    }
                    catch (e) {
                    }
                    if (extendedUnit) {
                        var m = this.pathMap(extendedUnit);
                        if (m) {
                            for (var _b = 0, _c = Object.keys(m); _b < _c.length; _b++) {
                                var k = _c[_b];
                                result[k] = m[k];
                            }
                        }
                    }
                }
            }
        }
        return result;
    };
    NamespaceResolver.prototype.lexLessEq = function (a, b) {
        if (a.length > b.length) {
            return false;
        }
        if (a.length < b.length) {
            return true;
        }
        for (var i = 0; i < a.length; i++) {
            var seg_a = a[i];
            var seg_b = b[i];
            if (seg_a < seg_b) {
                return true;
            }
            else if (seg_a > seg_b) {
                return false;
            }
        }
        return true;
    };
    NamespaceResolver.prototype.hasFragments = function (unit) {
        this.calculateExpandedNamespaces(unit);
        return this._hasFragments[unit.absolutePath()] ? true : false;
    };
    NamespaceResolver.prototype.unitModel = function (unit) {
        var aPath = unit.absolutePath();
        var result = this._unitModels[aPath];
        if (!result) {
            result = new UnitModel(unit);
            this._unitModels[aPath] = result;
        }
        return result;
    };
    return NamespaceResolver;
}());
exports.NamespaceResolver = NamespaceResolver;
var UsesInfo = (function () {
    function UsesInfo(usesNodes, unit, includePath) {
        this.usesNodes = usesNodes;
        this.unit = unit;
        this.includePath = includePath;
        this.namespaceSegments = this.usesNodes.map(function (x) { return x.key(); });
    }
    UsesInfo.prototype.steps = function () {
        return this.namespaceSegments.length;
    };
    UsesInfo.prototype.namespace = function () {
        return this.namespaceSegments.join(".");
    };
    UsesInfo.prototype.absolutePath = function () {
        return this.unit.absolutePath();
    };
    return UsesInfo;
}());
exports.UsesInfo = UsesInfo;
var ElementsCollection = (function () {
    function ElementsCollection(name) {
        this.name = name;
        this.array = [];
        this.map = {};
    }
    ElementsCollection.isInstance = function (instance) {
        return instance != null && instance.getClassIdentifier
            && typeof (instance.getClassIdentifier) == "function"
            && _.contains(instance.getClassIdentifier(), ElementsCollection.CLASS_IDENTIFIER);
    };
    ElementsCollection.prototype.getClassIdentifier = function () {
        var superIdentifiers = [];
        return superIdentifiers.concat(ElementsCollection.CLASS_IDENTIFIER);
    };
    ElementsCollection.prototype.addElement = function (node) {
        this.array.push(node);
        this.map[node.key()] = node;
    };
    ElementsCollection.prototype.hasElement = function (name) {
        return (this.map[name] != null);
    };
    ElementsCollection.prototype.getElement = function (name) {
        return this.map[name];
    };
    ElementsCollection.prototype.isEmpty = function () {
        return this.array.length == 0;
    };
    ElementsCollection.CLASS_IDENTIFIER = "namespaceResolver.ElementsCollection";
    return ElementsCollection;
}());
exports.ElementsCollection = ElementsCollection;
var UnitModel = (function () {
    function UnitModel(unit) {
        this.unit = unit;
        this.initCollections();
    }
    UnitModel.prototype.initCollections = function () {
        this.types = new ElementsCollection(def.universesInfo.Universe10.LibraryBase.properties.types.name);
        this.annotationTypes = new ElementsCollection(def.universesInfo.Universe10.LibraryBase.properties.annotationTypes.name);
        this.securitySchemes = new ElementsCollection(def.universesInfo.Universe10.LibraryBase.properties.securitySchemes.name);
        this.traits = new ElementsCollection(def.universesInfo.Universe10.LibraryBase.properties.traits.name);
        this.resourceTypes = new ElementsCollection(def.universesInfo.Universe10.LibraryBase.properties.resourceTypes.name);
        var rootNode = this.unit.ast();
        if (!rootNode) {
            return;
        }
        if (!this.isLibraryBaseDescendant(this.unit)) {
            return;
        }
        var isRAML08 = (hlImpl.ramlFirstLine(this.unit.contents())[1] == "0.8");
        for (var _i = 0, _a = rootNode.children(); _i < _a.length; _i++) {
            var ch = _a[_i];
            var key = ch.key();
            if (key == def.universesInfo.Universe10.LibraryBase.properties.types.name) {
                this.contributeCollection(this.types, ch.children());
            }
            else if (key == def.universesInfo.Universe10.LibraryBase.properties.annotationTypes.name) {
                this.contributeCollection(this.annotationTypes, ch.children());
            }
            else if (key == def.universesInfo.Universe10.LibraryBase.properties.securitySchemes.name) {
                this.contributeCollection(this.securitySchemes, ch.children(), isRAML08);
            }
            else if (key == def.universesInfo.Universe10.LibraryBase.properties.traits.name) {
                this.contributeCollection(this.traits, ch.children(), isRAML08);
            }
            else if (key == def.universesInfo.Universe10.LibraryBase.properties.resourceTypes.name) {
                this.contributeCollection(this.resourceTypes, ch.children(), isRAML08);
            }
        }
    };
    UnitModel.prototype.contributeCollection = function (c, nodes, isRAML08) {
        if (isRAML08 === void 0) { isRAML08 = false; }
        var _nodes;
        if (isRAML08) {
            _nodes = [];
            nodes.forEach(function (x) {
                var children = x.children();
                if (children.length > 0) {
                    _nodes.push(children[0]);
                }
            });
        }
        else {
            _nodes = nodes;
        }
        _nodes.forEach(function (x) { return c.addElement(x); });
    };
    UnitModel.prototype.isLibraryBaseDescendant = function (unit) {
        var _this = this;
        var fLine = hlImpl.ramlFirstLine(unit.contents());
        if (!fLine) {
            return false;
        }
        if (fLine.length < 3 || !fLine[2]) {
            return true;
        }
        var typeName = fLine[2];
        if (!this.libTypeDescendants) {
            this.libTypeDescendants = {};
            var u = def.getUniverse("RAML10");
            var libType = u.type(def.universesInfo.Universe10.LibraryBase.name);
            [libType].concat(libType.allSubTypes()).forEach(function (x) { return _this.libTypeDescendants[x.nameId()] = true; });
        }
        return this.libTypeDescendants[typeName];
    };
    return UnitModel;
}());
exports.UnitModel = UnitModel;
function extendsValue(u) {
    var node = u.ast();
    var eNode = _.find(node.children(), function (x) { return x.key() == universes.Universe10.Extension.properties.extends.name; });
    return eNode && eNode.value();
}
