// raml-1-0-parser/dist/raml1/artifacts/raml10parserapi.js
/**
 * <p>See <a href="http://raml.org">http://raml.org</a> for more information about RAML.</p>
 *
 * <p>This parser is at a beta state of development, as part of the API Workbench development cycle (<a href="http://apiworkbench.com">http://apiworkbench.com</a>).</p>
 *
 * <p><a href="https://github.com/raml-org/raml-js-parser-2/blob/master/documentation/GettingStarted.md">Getting Started Guide</a> describes the first steps with the parser.</p>
 *
 * <h2>Installation</h2>
 *
 * <pre><code>git clone https://github.com/raml-org/raml-js-parser-2
 *
 * cd raml-js-parser-2
 *
 * npm install
 *
 * node test/test.js  //here you should observe JSON representation of XKCD API in your console
 *
 * node test/testAsync.js  //same as above but in asynchronous mode
 * </code></pre>
 *
 * <h2>Usage</h2>
 *
 * <ul>
 * <li>For parser usage example refer to <code>test/test.js</code></li>
 * <li>For asynchrounous usage example refer to <code>test/testAsync.js</code></li>
 * </ul>
 **/
"use strict";
/**
 * Custom type guard for Api. Returns true if node is instance of Api. Returns false otherwise.
 * Also returns false for super interfaces of Api.
 */
function isApi(node) {
    return node.kind() == "Api" && node.RAMLVersion() == "RAML10";
}
exports.isApi = isApi;
/**
 * Custom type guard for LibraryBase. Returns true if node is instance of LibraryBase. Returns false otherwise.
 * Also returns false for super interfaces of LibraryBase.
 */
function isLibraryBase(node) {
    return node.kind() == "LibraryBase" && node.RAMLVersion() == "RAML10";
}
exports.isLibraryBase = isLibraryBase;
/**
 * Custom type guard for RAMLLanguageElement. Returns true if node is instance of RAMLLanguageElement. Returns false otherwise.
 * Also returns false for super interfaces of RAMLLanguageElement.
 */
function isRAMLLanguageElement(node) {
    return node.kind() == "RAMLLanguageElement" && node.RAMLVersion() == "RAML10";
}
exports.isRAMLLanguageElement = isRAMLLanguageElement;
/**
 * Custom type guard for MarkdownString. Returns true if node is instance of MarkdownString. Returns false otherwise.
 * Also returns false for super interfaces of MarkdownString.
 */
function isMarkdownString(node) {
    return node.kind() == "MarkdownString" && node.RAMLVersion() == "RAML10";
}
exports.isMarkdownString = isMarkdownString;
/**
 * Custom type guard for StringType. Returns true if node is instance of StringType. Returns false otherwise.
 * Also returns false for super interfaces of StringType.
 */
function isStringType(node) {
    return node.kind() == "StringType" && node.RAMLVersion() == "RAML10";
}
exports.isStringType = isStringType;
/**
 * Custom type guard for ValueType. Returns true if node is instance of ValueType. Returns false otherwise.
 * Also returns false for super interfaces of ValueType.
 */
function isValueType(node) {
    return node.kind() == "ValueType" && node.RAMLVersion() == "RAML10";
}
exports.isValueType = isValueType;
/**
 * Custom type guard for AnyType. Returns true if node is instance of AnyType. Returns false otherwise.
 * Also returns false for super interfaces of AnyType.
 */
function isAnyType(node) {
    return node.kind() == "AnyType" && node.RAMLVersion() == "RAML10";
}
exports.isAnyType = isAnyType;
/**
 * Custom type guard for NumberType. Returns true if node is instance of NumberType. Returns false otherwise.
 * Also returns false for super interfaces of NumberType.
 */
function isNumberType(node) {
    return node.kind() == "NumberType" && node.RAMLVersion() == "RAML10";
}
exports.isNumberType = isNumberType;
/**
 * Custom type guard for BooleanType. Returns true if node is instance of BooleanType. Returns false otherwise.
 * Also returns false for super interfaces of BooleanType.
 */
function isBooleanType(node) {
    return node.kind() == "BooleanType" && node.RAMLVersion() == "RAML10";
}
exports.isBooleanType = isBooleanType;
/**
 * Custom type guard for Reference. Returns true if node is instance of Reference. Returns false otherwise.
 * Also returns false for super interfaces of Reference.
 */
function isReference(node) {
    return node.kind() == "Reference" && node.RAMLVersion() == "RAML10";
}
exports.isReference = isReference;
/**
 * Custom type guard for AnnotationRef. Returns true if node is instance of AnnotationRef. Returns false otherwise.
 * Also returns false for super interfaces of AnnotationRef.
 */
function isAnnotationRef(node) {
    return node.kind() == "AnnotationRef" && node.RAMLVersion() == "RAML10";
}
exports.isAnnotationRef = isAnnotationRef;
/**
 * Custom type guard for AnnotationTypeDeclaration. Returns true if node is instance of AnnotationTypeDeclaration. Returns false otherwise.
 * Also returns false for super interfaces of AnnotationTypeDeclaration.
 */
function isAnnotationTypeDeclaration(node) {
    return node.kind() == "AnnotationTypeDeclaration" && node.RAMLVersion() == "RAML10";
}
exports.isAnnotationTypeDeclaration = isAnnotationTypeDeclaration;
/**
 * Custom type guard for TypeDeclaration. Returns true if node is instance of TypeDeclaration. Returns false otherwise.
 * Also returns false for super interfaces of TypeDeclaration.
 */
function isTypeDeclaration(node) {
    return node.kind() == "TypeDeclaration" && node.RAMLVersion() == "RAML10";
}
exports.isTypeDeclaration = isTypeDeclaration;
/**
 * Custom type guard for ModelLocation. Returns true if node is instance of ModelLocation. Returns false otherwise.
 * Also returns false for super interfaces of ModelLocation.
 */
function isModelLocation(node) {
    return node.kind() == "ModelLocation" && node.RAMLVersion() == "RAML10";
}
exports.isModelLocation = isModelLocation;
/**
 * Custom type guard for LocationKind. Returns true if node is instance of LocationKind. Returns false otherwise.
 * Also returns false for super interfaces of LocationKind.
 */
function isLocationKind(node) {
    return node.kind() == "LocationKind" && node.RAMLVersion() == "RAML10";
}
exports.isLocationKind = isLocationKind;
/**
 * Custom type guard for ExampleSpec. Returns true if node is instance of ExampleSpec. Returns false otherwise.
 * Also returns false for super interfaces of ExampleSpec.
 */
function isExampleSpec(node) {
    return node.kind() == "ExampleSpec" && node.RAMLVersion() == "RAML10";
}
exports.isExampleSpec = isExampleSpec;
/**
 * Custom type guard for ArrayTypeDeclaration. Returns true if node is instance of ArrayTypeDeclaration. Returns false otherwise.
 * Also returns false for super interfaces of ArrayTypeDeclaration.
 */
function isArrayTypeDeclaration(node) {
    return node.kind() == "ArrayTypeDeclaration" && node.RAMLVersion() == "RAML10";
}
exports.isArrayTypeDeclaration = isArrayTypeDeclaration;
/**
 * Custom type guard for ArrayAnnotationTypeDeclaration. Returns true if node is instance of ArrayAnnotationTypeDeclaration. Returns false otherwise.
 * Also returns false for super interfaces of ArrayAnnotationTypeDeclaration.
 */
function isArrayAnnotationTypeDeclaration(node) {
    return node.kind() == "ArrayAnnotationTypeDeclaration" && node.RAMLVersion() == "RAML10";
}
exports.isArrayAnnotationTypeDeclaration = isArrayAnnotationTypeDeclaration;
/**
 * Custom type guard for AnnotationTarget. Returns true if node is instance of AnnotationTarget. Returns false otherwise.
 * Also returns false for super interfaces of AnnotationTarget.
 */
function isAnnotationTarget(node) {
    return node.kind() == "AnnotationTarget" && node.RAMLVersion() == "RAML10";
}
exports.isAnnotationTarget = isAnnotationTarget;
/**
 * Custom type guard for UnionTypeDeclaration. Returns true if node is instance of UnionTypeDeclaration. Returns false otherwise.
 * Also returns false for super interfaces of UnionTypeDeclaration.
 */
function isUnionTypeDeclaration(node) {
    return node.kind() == "UnionTypeDeclaration" && node.RAMLVersion() == "RAML10";
}
exports.isUnionTypeDeclaration = isUnionTypeDeclaration;
/**
 * Custom type guard for UnionAnnotationTypeDeclaration. Returns true if node is instance of UnionAnnotationTypeDeclaration. Returns false otherwise.
 * Also returns false for super interfaces of UnionAnnotationTypeDeclaration.
 */
function isUnionAnnotationTypeDeclaration(node) {
    return node.kind() == "UnionAnnotationTypeDeclaration" && node.RAMLVersion() == "RAML10";
}
exports.isUnionAnnotationTypeDeclaration = isUnionAnnotationTypeDeclaration;
/**
 * Custom type guard for ObjectTypeDeclaration. Returns true if node is instance of ObjectTypeDeclaration. Returns false otherwise.
 * Also returns false for super interfaces of ObjectTypeDeclaration.
 */
function isObjectTypeDeclaration(node) {
    return node.kind() == "ObjectTypeDeclaration" && node.RAMLVersion() == "RAML10";
}
exports.isObjectTypeDeclaration = isObjectTypeDeclaration;
/**
 * Custom type guard for ObjectAnnotationTypeDeclaration. Returns true if node is instance of ObjectAnnotationTypeDeclaration. Returns false otherwise.
 * Also returns false for super interfaces of ObjectAnnotationTypeDeclaration.
 */
function isObjectAnnotationTypeDeclaration(node) {
    return node.kind() == "ObjectAnnotationTypeDeclaration" && node.RAMLVersion() == "RAML10";
}
exports.isObjectAnnotationTypeDeclaration = isObjectAnnotationTypeDeclaration;
/**
 * Custom type guard for StringTypeDeclaration. Returns true if node is instance of StringTypeDeclaration. Returns false otherwise.
 * Also returns false for super interfaces of StringTypeDeclaration.
 */
function isStringTypeDeclaration(node) {
    return node.kind() == "StringTypeDeclaration" && node.RAMLVersion() == "RAML10";
}
exports.isStringTypeDeclaration = isStringTypeDeclaration;
/**
 * Custom type guard for StringAnnotationTypeDeclaration. Returns true if node is instance of StringAnnotationTypeDeclaration. Returns false otherwise.
 * Also returns false for super interfaces of StringAnnotationTypeDeclaration.
 */
function isStringAnnotationTypeDeclaration(node) {
    return node.kind() == "StringAnnotationTypeDeclaration" && node.RAMLVersion() == "RAML10";
}
exports.isStringAnnotationTypeDeclaration = isStringAnnotationTypeDeclaration;
/**
 * Custom type guard for BooleanTypeDeclaration. Returns true if node is instance of BooleanTypeDeclaration. Returns false otherwise.
 * Also returns false for super interfaces of BooleanTypeDeclaration.
 */
function isBooleanTypeDeclaration(node) {
    return node.kind() == "BooleanTypeDeclaration" && node.RAMLVersion() == "RAML10";
}
exports.isBooleanTypeDeclaration = isBooleanTypeDeclaration;
/**
 * Custom type guard for BooleanAnnotationTypeDeclaration. Returns true if node is instance of BooleanAnnotationTypeDeclaration. Returns false otherwise.
 * Also returns false for super interfaces of BooleanAnnotationTypeDeclaration.
 */
function isBooleanAnnotationTypeDeclaration(node) {
    return node.kind() == "BooleanAnnotationTypeDeclaration" && node.RAMLVersion() == "RAML10";
}
exports.isBooleanAnnotationTypeDeclaration = isBooleanAnnotationTypeDeclaration;
/**
 * Custom type guard for NumberTypeDeclaration. Returns true if node is instance of NumberTypeDeclaration. Returns false otherwise.
 * Also returns false for super interfaces of NumberTypeDeclaration.
 */
function isNumberTypeDeclaration(node) {
    return node.kind() == "NumberTypeDeclaration" && node.RAMLVersion() == "RAML10";
}
exports.isNumberTypeDeclaration = isNumberTypeDeclaration;
/**
 * Custom type guard for IntegerTypeDeclaration. Returns true if node is instance of IntegerTypeDeclaration. Returns false otherwise.
 * Also returns false for super interfaces of IntegerTypeDeclaration.
 */
function isIntegerTypeDeclaration(node) {
    return node.kind() == "IntegerTypeDeclaration" && node.RAMLVersion() == "RAML10";
}
exports.isIntegerTypeDeclaration = isIntegerTypeDeclaration;
/**
 * Custom type guard for NumberAnnotationTypeDeclaration. Returns true if node is instance of NumberAnnotationTypeDeclaration. Returns false otherwise.
 * Also returns false for super interfaces of NumberAnnotationTypeDeclaration.
 */
function isNumberAnnotationTypeDeclaration(node) {
    return node.kind() == "NumberAnnotationTypeDeclaration" && node.RAMLVersion() == "RAML10";
}
exports.isNumberAnnotationTypeDeclaration = isNumberAnnotationTypeDeclaration;
/**
 * Custom type guard for DateTypeDeclaration. Returns true if node is instance of DateTypeDeclaration. Returns false otherwise.
 * Also returns false for super interfaces of DateTypeDeclaration.
 */
function isDateTypeDeclaration(node) {
    return node.kind() == "DateTypeDeclaration" && node.RAMLVersion() == "RAML10";
}
exports.isDateTypeDeclaration = isDateTypeDeclaration;
/**
 * Custom type guard for DateTypeAnnotationDeclaration. Returns true if node is instance of DateTypeAnnotationDeclaration. Returns false otherwise.
 * Also returns false for super interfaces of DateTypeAnnotationDeclaration.
 */
function isDateTypeAnnotationDeclaration(node) {
    return node.kind() == "DateTypeAnnotationDeclaration" && node.RAMLVersion() == "RAML10";
}
exports.isDateTypeAnnotationDeclaration = isDateTypeAnnotationDeclaration;
/**
 * Custom type guard for FileTypeDeclaration. Returns true if node is instance of FileTypeDeclaration. Returns false otherwise.
 * Also returns false for super interfaces of FileTypeDeclaration.
 */
function isFileTypeDeclaration(node) {
    return node.kind() == "FileTypeDeclaration" && node.RAMLVersion() == "RAML10";
}
exports.isFileTypeDeclaration = isFileTypeDeclaration;
/**
 * Custom type guard for ContentType. Returns true if node is instance of ContentType. Returns false otherwise.
 * Also returns false for super interfaces of ContentType.
 */
function isContentType(node) {
    return node.kind() == "ContentType" && node.RAMLVersion() == "RAML10";
}
exports.isContentType = isContentType;
/**
 * Custom type guard for SecuritySchemeRef. Returns true if node is instance of SecuritySchemeRef. Returns false otherwise.
 * Also returns false for super interfaces of SecuritySchemeRef.
 */
function isSecuritySchemeRef(node) {
    return node.kind() == "SecuritySchemeRef" && node.RAMLVersion() == "RAML10";
}
exports.isSecuritySchemeRef = isSecuritySchemeRef;
/**
 * Custom type guard for AbstractSecurityScheme. Returns true if node is instance of AbstractSecurityScheme. Returns false otherwise.
 * Also returns false for super interfaces of AbstractSecurityScheme.
 */
function isAbstractSecurityScheme(node) {
    return node.kind() == "AbstractSecurityScheme" && node.RAMLVersion() == "RAML10";
}
exports.isAbstractSecurityScheme = isAbstractSecurityScheme;
/**
 * Custom type guard for SecuritySchemePart. Returns true if node is instance of SecuritySchemePart. Returns false otherwise.
 * Also returns false for super interfaces of SecuritySchemePart.
 */
function isSecuritySchemePart(node) {
    return node.kind() == "SecuritySchemePart" && node.RAMLVersion() == "RAML10";
}
exports.isSecuritySchemePart = isSecuritySchemePart;
/**
 * Custom type guard for MethodBase. Returns true if node is instance of MethodBase. Returns false otherwise.
 * Also returns false for super interfaces of MethodBase.
 */
function isMethodBase(node) {
    return node.kind() == "MethodBase" && node.RAMLVersion() == "RAML10";
}
exports.isMethodBase = isMethodBase;
/**
 * Custom type guard for HasNormalParameters. Returns true if node is instance of HasNormalParameters. Returns false otherwise.
 * Also returns false for super interfaces of HasNormalParameters.
 */
function isHasNormalParameters(node) {
    return node.kind() == "HasNormalParameters" && node.RAMLVersion() == "RAML10";
}
exports.isHasNormalParameters = isHasNormalParameters;
/**
 * Custom type guard for Response. Returns true if node is instance of Response. Returns false otherwise.
 * Also returns false for super interfaces of Response.
 */
function isResponse(node) {
    return node.kind() == "Response" && node.RAMLVersion() == "RAML10";
}
exports.isResponse = isResponse;
/**
 * Custom type guard for StatusCodeString. Returns true if node is instance of StatusCodeString. Returns false otherwise.
 * Also returns false for super interfaces of StatusCodeString.
 */
function isStatusCodeString(node) {
    return node.kind() == "StatusCodeString" && node.RAMLVersion() == "RAML10";
}
exports.isStatusCodeString = isStatusCodeString;
/**
 * Custom type guard for TraitRef. Returns true if node is instance of TraitRef. Returns false otherwise.
 * Also returns false for super interfaces of TraitRef.
 */
function isTraitRef(node) {
    return node.kind() == "TraitRef" && node.RAMLVersion() == "RAML10";
}
exports.isTraitRef = isTraitRef;
/**
 * Custom type guard for Trait. Returns true if node is instance of Trait. Returns false otherwise.
 * Also returns false for super interfaces of Trait.
 */
function isTrait(node) {
    return node.kind() == "Trait" && node.RAMLVersion() == "RAML10";
}
exports.isTrait = isTrait;
/**
 * Custom type guard for Library. Returns true if node is instance of Library. Returns false otherwise.
 * Also returns false for super interfaces of Library.
 */
function isLibrary(node) {
    return node.kind() == "Library" && node.RAMLVersion() == "RAML10";
}
exports.isLibrary = isLibrary;
/**
 * Custom type guard for Method. Returns true if node is instance of Method. Returns false otherwise.
 * Also returns false for super interfaces of Method.
 */
function isMethod(node) {
    return node.kind() == "Method" && node.RAMLVersion() == "RAML10";
}
exports.isMethod = isMethod;
/**
 * Custom type guard for SecuritySchemeSettings. Returns true if node is instance of SecuritySchemeSettings. Returns false otherwise.
 * Also returns false for super interfaces of SecuritySchemeSettings.
 */
function isSecuritySchemeSettings(node) {
    return node.kind() == "SecuritySchemeSettings" && node.RAMLVersion() == "RAML10";
}
exports.isSecuritySchemeSettings = isSecuritySchemeSettings;
/**
 * Custom type guard for OAuth1SecuritySchemeSettings. Returns true if node is instance of OAuth1SecuritySchemeSettings. Returns false otherwise.
 * Also returns false for super interfaces of OAuth1SecuritySchemeSettings.
 */
function isOAuth1SecuritySchemeSettings(node) {
    return node.kind() == "OAuth1SecuritySchemeSettings" && node.RAMLVersion() == "RAML10";
}
exports.isOAuth1SecuritySchemeSettings = isOAuth1SecuritySchemeSettings;
/**
 * Custom type guard for FixedUriString. Returns true if node is instance of FixedUriString. Returns false otherwise.
 * Also returns false for super interfaces of FixedUriString.
 */
function isFixedUriString(node) {
    return node.kind() == "FixedUriString" && node.RAMLVersion() == "RAML10";
}
exports.isFixedUriString = isFixedUriString;
/**
 * Custom type guard for OAuth2SecuritySchemeSettings. Returns true if node is instance of OAuth2SecuritySchemeSettings. Returns false otherwise.
 * Also returns false for super interfaces of OAuth2SecuritySchemeSettings.
 */
function isOAuth2SecuritySchemeSettings(node) {
    return node.kind() == "OAuth2SecuritySchemeSettings" && node.RAMLVersion() == "RAML10";
}
exports.isOAuth2SecuritySchemeSettings = isOAuth2SecuritySchemeSettings;
/**
 * Custom type guard for PassThroughSecuritySchemeSettings. Returns true if node is instance of PassThroughSecuritySchemeSettings. Returns false otherwise.
 * Also returns false for super interfaces of PassThroughSecuritySchemeSettings.
 */
function isPassThroughSecuritySchemeSettings(node) {
    return node.kind() == "PassThroughSecuritySchemeSettings" && node.RAMLVersion() == "RAML10";
}
exports.isPassThroughSecuritySchemeSettings = isPassThroughSecuritySchemeSettings;
/**
 * Custom type guard for OAuth2SecurityScheme. Returns true if node is instance of OAuth2SecurityScheme. Returns false otherwise.
 * Also returns false for super interfaces of OAuth2SecurityScheme.
 */
function isOAuth2SecurityScheme(node) {
    return node.kind() == "OAuth2SecurityScheme" && node.RAMLVersion() == "RAML10";
}
exports.isOAuth2SecurityScheme = isOAuth2SecurityScheme;
/**
 * Custom type guard for OAuth1SecurityScheme. Returns true if node is instance of OAuth1SecurityScheme. Returns false otherwise.
 * Also returns false for super interfaces of OAuth1SecurityScheme.
 */
function isOAuth1SecurityScheme(node) {
    return node.kind() == "OAuth1SecurityScheme" && node.RAMLVersion() == "RAML10";
}
exports.isOAuth1SecurityScheme = isOAuth1SecurityScheme;
/**
 * Custom type guard for PassThroughSecurityScheme. Returns true if node is instance of PassThroughSecurityScheme. Returns false otherwise.
 * Also returns false for super interfaces of PassThroughSecurityScheme.
 */
function isPassThroughSecurityScheme(node) {
    return node.kind() == "PassThroughSecurityScheme" && node.RAMLVersion() == "RAML10";
}
exports.isPassThroughSecurityScheme = isPassThroughSecurityScheme;
/**
 * Custom type guard for BasicSecurityScheme. Returns true if node is instance of BasicSecurityScheme. Returns false otherwise.
 * Also returns false for super interfaces of BasicSecurityScheme.
 */
function isBasicSecurityScheme(node) {
    return node.kind() == "BasicSecurityScheme" && node.RAMLVersion() == "RAML10";
}
exports.isBasicSecurityScheme = isBasicSecurityScheme;
/**
 * Custom type guard for DigestSecurityScheme. Returns true if node is instance of DigestSecurityScheme. Returns false otherwise.
 * Also returns false for super interfaces of DigestSecurityScheme.
 */
function isDigestSecurityScheme(node) {
    return node.kind() == "DigestSecurityScheme" && node.RAMLVersion() == "RAML10";
}
exports.isDigestSecurityScheme = isDigestSecurityScheme;
/**
 * Custom type guard for CustomSecurityScheme. Returns true if node is instance of CustomSecurityScheme. Returns false otherwise.
 * Also returns false for super interfaces of CustomSecurityScheme.
 */
function isCustomSecurityScheme(node) {
    return node.kind() == "CustomSecurityScheme" && node.RAMLVersion() == "RAML10";
}
exports.isCustomSecurityScheme = isCustomSecurityScheme;
/**
 * Custom type guard for ResourceTypeRef. Returns true if node is instance of ResourceTypeRef. Returns false otherwise.
 * Also returns false for super interfaces of ResourceTypeRef.
 */
function isResourceTypeRef(node) {
    return node.kind() == "ResourceTypeRef" && node.RAMLVersion() == "RAML10";
}
exports.isResourceTypeRef = isResourceTypeRef;
/**
 * Custom type guard for ResourceType. Returns true if node is instance of ResourceType. Returns false otherwise.
 * Also returns false for super interfaces of ResourceType.
 */
function isResourceType(node) {
    return node.kind() == "ResourceType" && node.RAMLVersion() == "RAML10";
}
exports.isResourceType = isResourceType;
/**
 * Custom type guard for ResourceBase. Returns true if node is instance of ResourceBase. Returns false otherwise.
 * Also returns false for super interfaces of ResourceBase.
 */
function isResourceBase(node) {
    return node.kind() == "ResourceBase" && node.RAMLVersion() == "RAML10";
}
exports.isResourceBase = isResourceBase;
/**
 * Custom type guard for Resource. Returns true if node is instance of Resource. Returns false otherwise.
 * Also returns false for super interfaces of Resource.
 */
function isResource(node) {
    return node.kind() == "Resource" && node.RAMLVersion() == "RAML10";
}
exports.isResource = isResource;
/**
 * Custom type guard for RelativeUriString. Returns true if node is instance of RelativeUriString. Returns false otherwise.
 * Also returns false for super interfaces of RelativeUriString.
 */
function isRelativeUriString(node) {
    return node.kind() == "RelativeUriString" && node.RAMLVersion() == "RAML10";
}
exports.isRelativeUriString = isRelativeUriString;
/**
 * Custom type guard for UriTemplate. Returns true if node is instance of UriTemplate. Returns false otherwise.
 * Also returns false for super interfaces of UriTemplate.
 */
function isUriTemplate(node) {
    return node.kind() == "UriTemplate" && node.RAMLVersion() == "RAML10";
}
exports.isUriTemplate = isUriTemplate;
/**
 * Custom type guard for FullUriTemplateString. Returns true if node is instance of FullUriTemplateString. Returns false otherwise.
 * Also returns false for super interfaces of FullUriTemplateString.
 */
function isFullUriTemplateString(node) {
    return node.kind() == "FullUriTemplateString" && node.RAMLVersion() == "RAML10";
}
exports.isFullUriTemplateString = isFullUriTemplateString;
/**
 * Custom type guard for SchemaString. Returns true if node is instance of SchemaString. Returns false otherwise.
 * Also returns false for super interfaces of SchemaString.
 */
function isSchemaString(node) {
    return node.kind() == "SchemaString" && node.RAMLVersion() == "RAML10";
}
exports.isSchemaString = isSchemaString;
/**
 * Custom type guard for JSonSchemaString. Returns true if node is instance of JSonSchemaString. Returns false otherwise.
 * Also returns false for super interfaces of JSonSchemaString.
 */
function isJSonSchemaString(node) {
    return node.kind() == "JSonSchemaString" && node.RAMLVersion() == "RAML10";
}
exports.isJSonSchemaString = isJSonSchemaString;
/**
 * Custom type guard for XMLSchemaString. Returns true if node is instance of XMLSchemaString. Returns false otherwise.
 * Also returns false for super interfaces of XMLSchemaString.
 */
function isXMLSchemaString(node) {
    return node.kind() == "XMLSchemaString" && node.RAMLVersion() == "RAML10";
}
exports.isXMLSchemaString = isXMLSchemaString;
/**
 * Custom type guard for ExampleString. Returns true if node is instance of ExampleString. Returns false otherwise.
 * Also returns false for super interfaces of ExampleString.
 */
function isExampleString(node) {
    return node.kind() == "ExampleString" && node.RAMLVersion() == "RAML10";
}
exports.isExampleString = isExampleString;
/**
 * Custom type guard for MimeType. Returns true if node is instance of MimeType. Returns false otherwise.
 * Also returns false for super interfaces of MimeType.
 */
function isMimeType(node) {
    return node.kind() == "MimeType" && node.RAMLVersion() == "RAML10";
}
exports.isMimeType = isMimeType;
/**
 * Custom type guard for DocumentationItem. Returns true if node is instance of DocumentationItem. Returns false otherwise.
 * Also returns false for super interfaces of DocumentationItem.
 */
function isDocumentationItem(node) {
    return node.kind() == "DocumentationItem" && node.RAMLVersion() == "RAML10";
}
exports.isDocumentationItem = isDocumentationItem;
/**
 * Custom type guard for GlobalSchema. Returns true if node is instance of GlobalSchema. Returns false otherwise.
 * Also returns false for super interfaces of GlobalSchema.
 */
function isGlobalSchema(node) {
    return node.kind() == "GlobalSchema" && node.RAMLVersion() == "RAML10";
}
exports.isGlobalSchema = isGlobalSchema;
/**
 * Custom type guard for RAMLSimpleElement. Returns true if node is instance of RAMLSimpleElement. Returns false otherwise.
 * Also returns false for super interfaces of RAMLSimpleElement.
 */
function isRAMLSimpleElement(node) {
    return node.kind() == "RAMLSimpleElement" && node.RAMLVersion() == "RAML10";
}
exports.isRAMLSimpleElement = isRAMLSimpleElement;
/**
 * Custom type guard for ImportDeclaration. Returns true if node is instance of ImportDeclaration. Returns false otherwise.
 * Also returns false for super interfaces of ImportDeclaration.
 */
function isImportDeclaration(node) {
    return node.kind() == "ImportDeclaration" && node.RAMLVersion() == "RAML10";
}
exports.isImportDeclaration = isImportDeclaration;
/**
 * Custom type guard for Overlay. Returns true if node is instance of Overlay. Returns false otherwise.
 * Also returns false for super interfaces of Overlay.
 */
function isOverlay(node) {
    return node.kind() == "Overlay" && node.RAMLVersion() == "RAML10";
}
exports.isOverlay = isOverlay;
/**
 * Custom type guard for Extension. Returns true if node is instance of Extension. Returns false otherwise.
 * Also returns false for super interfaces of Extension.
 */
function isExtension(node) {
    return node.kind() == "Extension" && node.RAMLVersion() == "RAML10";
}
exports.isExtension = isExtension;
