// ts-structure-parser/dist/index.js
"use strict";
exports.helpers = require("./helpers");
var tsStructureParser = require("./tsStructureParser");
var EnumDeclaration = (function () {
    function EnumDeclaration() {
    }
    return EnumDeclaration;
}());
exports.EnumDeclaration = EnumDeclaration;
(function (TypeKind) {
    TypeKind[TypeKind["BASIC"] = 0] = "BASIC";
    TypeKind[TypeKind["ARRAY"] = 1] = "ARRAY";
    TypeKind[TypeKind["UNION"] = 2] = "UNION";
})(exports.TypeKind || (exports.TypeKind = {}));
var TypeKind = exports.TypeKind;
function classDecl(name, isInteface) {
    return {
        name: name,
        methods: [],
        typeParameters: [],
        typeParameterConstraint: [],
        implements: [],
        fields: [],
        isInterface: isInteface,
        annotations: [],
        extends: [],
        moduleName: null,
        annotationOverridings: {}
    };
}
exports.classDecl = classDecl;
function parseStruct(content, modules, mpth) {
    return tsStructureParser.parseStruct(content, modules, mpth);
}
exports.parseStruct = parseStruct;
