// ts-model/dist/TSDeclModel.js
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
exports.tsutil = require("./tsutil");
var _ = require("./underscore");
exports.MODEL_CLASS_MODEL_ELEMENT = '$resource-model-element';
exports.MODEL_CLASS_TYPE_DECLARATION = '$type-declaration';
exports.MODEL_CLASS_INTERFACE = '$interface-declaration';
exports.MODEL_CLASS_CLASS_DECLARATION = '$class-declaration';
exports.MODEL_CLASS_ANNOTATION_DECLARATION = '$annotation-declaration';
exports.MODEL_CLASS_ENUM_DECLARATION = '$enum-declaration';
exports.MODEL_CLASS_TYPE_ASSERTION = '$type-assertion';
exports.MODEL_CLASS_API_MODULE = '$api-module';
exports.MODEL_CLASS_UNIVERSE = '$universe';
exports.MODEL_CLASS_MEMBER = '$member';
exports.MODEL_CLASS_UNION_TYPE_REFERENCE = '$union-type-reference';
exports.MODEL_CLASS_SIMPLE_TYPE_REFERENCE = '$simple-type-reference';
exports.MODEL_CLASS_FUNCTION_REFERENCE = '$function-reference';
exports.MODEL_CLASS_ARRAY_REFERENCE = '$array-reference';
exports.MODEL_CLASS_DECLARED_INTERFACE_REFERENCE = '$declared-interface-reference';
exports.MODEL_CLASS_ANY_TYPE_REFERENCE = '$any-type-reference';
exports.MODEL_CLASS_STRUCTURAL_TYPE_REFERENCE = '$structural-type-reference';
exports.MODEL_CLASS_PARAM = '$param';
exports.MODEL_CLASS_STRING_VALUE = '$string-value';
exports.MODEL_CLASS_ARRAY_VALUE = '$array-value';
exports.MODEL_CLASS_API_ELEMENT_DECLARATION = '$api-element-declaration';
exports.MODEL_CLASS_CONSTRUCTOR = '$constructor';
//TODO HIDE Fields from unmanagable modification
//TODO Refine type decl type ref hieararchy a bit more
//TODO add classes, generics, metadata
var TSModelElement = (function () {
    function TSModelElement(parent, config) {
        if (parent === void 0) { parent = exports.Universe; }
        this.meta = {};
        this._annotations = [];
        this.setParent(parent);
        if (config)
            this._config = config;
        this._children = [];
    }
    TSModelElement.prototype.annotations = function () {
        return this._annotations;
    };
    TSModelElement.prototype.patchParent = function (parent) {
        this._parent = parent; //FIXME
    };
    TSModelElement.prototype.isEmpty = function () {
        return this._children.length == 0;
    };
    TSModelElement.prototype.parent = function () { return this._parent; };
    TSModelElement.prototype.children = function () {
        return this._children;
    };
    TSModelElement.prototype.comment = function () { return this._comment; };
    TSModelElement.prototype.root = function () {
        if (this._parent == exports.Universe) {
            return this;
        }
        return this._parent.root();
    };
    TSModelElement.prototype.setParent = function (parent) {
        if (!parent)
            return;
        this._parent = parent;
        this._config = parent._config;
        this._parent.addChild(this);
    };
    TSModelElement.prototype.removeChild = function (child) {
        if (child.parent() == this) {
            this._children = this._children.filter(function (x) { return x != child; });
        }
        child.patchParent(exports.Universe);
    };
    TSModelElement.prototype.addChild = function (child) {
        if (child.parent()) {
            child.parent().removeChild(child);
        }
        child.patchParent(this);
        this._children.push(child);
    };
    TSModelElement.prototype.serializeToString = function (isImpl) {
        if (isImpl === void 0) { isImpl = false; }
        throw new Error("You should override serialize to string always");
    };
    TSModelElement.prototype.modelClass = function () {
        return exports.MODEL_CLASS_MODEL_ELEMENT;
    };
    return TSModelElement;
}());
exports.TSModelElement = TSModelElement;
//TODO It should become an interface
var TSTypeDeclaration = (function (_super) {
    __extends(TSTypeDeclaration, _super);
    function TSTypeDeclaration(parent) {
        var _this = this;
        if (parent === void 0) { parent = null; }
        _super.call(this, parent);
        this.canBeOmmited = function () { return _this.locked ? false : _this.children().every(function (x) { return x.optional; }); };
        this.locked = false;
        this.extras = [""];
    }
    TSTypeDeclaration.prototype.addCode = function (code) {
        this.extras.push(code);
    };
    TSTypeDeclaration.prototype.toReference = function () { throw new Error("Implement in subclasses"); };
    TSTypeDeclaration.prototype.hash = function () { return this.serializeToString(); };
    TSTypeDeclaration.prototype.isFunctor = function () {
        return this.children().some(function (x) { return x.isAnonymousFunction(); });
    };
    TSTypeDeclaration.prototype.getFunctor = function () {
        return _.find(this.children(), function (x) { return x.isAnonymousFunction(); });
    };
    TSTypeDeclaration.prototype.visit = function (v) {
        if (v.startTypeDeclaration(this)) {
            this.children().forEach(function (x, i, arr) {
                x.visit(v);
                if (i != arr.length - 1)
                    v.betweenElements();
            });
            v.endTypeDeclaration(this);
        }
    };
    TSTypeDeclaration.prototype.modelClass = function () {
        return exports.MODEL_CLASS_TYPE_DECLARATION;
    };
    return TSTypeDeclaration;
}(TSModelElement));
exports.TSTypeDeclaration = TSTypeDeclaration;
var TSInterface = (function (_super) {
    __extends(TSInterface, _super);
    function TSInterface(p, name) {
        _super.call(this, p);
        this.extends = [];
        this.implements = [];
        this.name = name;
    }
    TSInterface.prototype.typeSignature = function () {
        var modelName = this.name;
        if (this.typeParameters && this.typeParameters.length > 0) {
            modelName += "<" + this.typeParameters.join(',') + ">";
        }
        return modelName;
    };
    TSInterface.prototype.hash = function () {
        var _this = this;
        return this.children().filter(function (x) { return !x.isPrivate; }).map(function (x) { return "\n" + x.serializeToString(_this.isImpl()) + "\n"; }).join('');
    };
    TSInterface.prototype.toReference = function () { return new TSDeclaredInterfaceReference(exports.Universe, this.name, this); };
    TSInterface.prototype.decl = function () {
        return "interface";
    };
    TSInterface.prototype.serializeToString = function () {
        var body = this.hash();
        return this.insertComment() + "export " + this.decl() + " " + this.name.concat(this.extendsString() + this.implementsString()) +
            "{" + this.extras.join("\n") + body + "}\n";
    };
    TSInterface.prototype.extendsString = function () {
        if (this.extends.length > 0) {
            return " extends " + this.extends.map(function (x) { return x.serializeToString(); }).join(",");
        }
        return "";
    };
    TSInterface.prototype.implementsString = function () {
        if (this.implements.length > 0) {
            return " implements " + this.implements.map(function (x) { return x.serializeToString(); }).join(",");
        }
        return "";
    };
    TSInterface.prototype.insertComment = function () {
        if (!this._comment) {
            return '';
        }
        return formattedComment(this._comment, 0);
    };
    TSInterface.prototype.modelClass = function () {
        return exports.MODEL_CLASS_INTERFACE;
    };
    TSInterface.prototype.isImpl = function () {
        return false;
    };
    return TSInterface;
}(TSTypeDeclaration));
exports.TSInterface = TSInterface;
//TODO INCORRECT INHERITANCE CHAIN
var TSClassDecl = (function (_super) {
    __extends(TSClassDecl, _super);
    function TSClassDecl() {
        _super.apply(this, arguments);
    }
    TSClassDecl.prototype.decl = function () {
        return "class";
    };
    TSClassDecl.prototype.modelClass = function () {
        return exports.MODEL_CLASS_CLASS_DECLARATION;
    };
    TSClassDecl.prototype.isImpl = function () {
        return true;
    };
    return TSClassDecl;
}(TSInterface));
exports.TSClassDecl = TSClassDecl;
var TSAnnotationDecl = (function (_super) {
    __extends(TSAnnotationDecl, _super);
    function TSAnnotationDecl() {
        _super.apply(this, arguments);
    }
    TSAnnotationDecl.prototype.decl = function () {
        return "annotation";
    };
    TSAnnotationDecl.prototype.modelClass = function () {
        return exports.MODEL_CLASS_ANNOTATION_DECLARATION;
    };
    TSAnnotationDecl.prototype.toReference = function () { return new TSDeclaredAnnotationReference(exports.Universe, this.name, this); };
    return TSAnnotationDecl;
}(TSInterface));
exports.TSAnnotationDecl = TSAnnotationDecl;
var TSEnumDecl = (function (_super) {
    __extends(TSEnumDecl, _super);
    function TSEnumDecl() {
        _super.apply(this, arguments);
    }
    TSEnumDecl.prototype.decl = function () {
        return "enum";
    };
    TSEnumDecl.prototype.modelClass = function () {
        return exports.MODEL_CLASS_ENUM_DECLARATION;
    };
    return TSEnumDecl;
}(TSInterface));
exports.TSEnumDecl = TSEnumDecl;
var TSTypeAssertion = (function (_super) {
    __extends(TSTypeAssertion, _super);
    function TSTypeAssertion(p, _name, _ref) {
        _super.call(this, p);
        this._name = _name;
        this._ref = _ref;
    }
    TSTypeAssertion.prototype.toReference = function () { return new TSSimpleTypeReference(exports.Universe, this._name); };
    TSTypeAssertion.prototype.serializeToString = function () {
        return "export type " + this._name + "=" + this._ref.serializeToString() + "\n";
    };
    TSTypeAssertion.prototype.ref = function () { return this._ref; };
    TSTypeAssertion.prototype.name = function () { return this._name; };
    TSTypeAssertion.prototype.modelClass = function () {
        return exports.MODEL_CLASS_TYPE_ASSERTION;
    };
    return TSTypeAssertion;
}(TSTypeDeclaration));
exports.TSTypeAssertion = TSTypeAssertion;
var TSUniverse = (function (_super) {
    __extends(TSUniverse, _super);
    function TSUniverse() {
        // super(this);
        //commented this call out as the new version of TS compiler does not eat this. For a reason.
        //so now first calling super with a fake parent
        _super.call(this, new TSModelElement());
        //and now setting a real parent as self
        //(not sure why we do this originally, I would put null here, creating a cycle doesnt look good)
        this.setParent(this);
    }
    TSUniverse.prototype.addChild = function (child) {
    };
    TSUniverse.prototype.setConfig = function (cfg) { this._config = cfg; };
    TSUniverse.prototype.getConfig = function () { return this._config; };
    TSUniverse.prototype.modelClass = function () {
        return exports.MODEL_CLASS_UNIVERSE;
    };
    return TSUniverse;
}(TSModelElement));
exports.TSUniverse = TSUniverse;
exports.Universe = new TSUniverse();
var TSAPIModule = (function (_super) {
    __extends(TSAPIModule, _super);
    function TSAPIModule() {
        _super.apply(this, arguments);
    }
    TSAPIModule.prototype.getInterface = function (nm) {
        return _.find(this.children(), function (x) { return x.name == nm; });
    };
    TSAPIModule.prototype.serializeToString = function () {
        var typeMap = {};
        this.children().forEach(function (x) { return typeMap[x.name] = x; });
        var covered = {};
        var sorted = [];
        var append = function (t) {
            if (covered[t.name]) {
                return;
            }
            covered[t.name] = true;
            var refs = t.extends;
            refs.forEach(function (ref) {
                if (ref instanceof TSSimpleTypeReference) {
                    var name = ref.name;
                    var st = typeMap[name];
                    if (st) {
                        append(st);
                    }
                }
            });
            sorted.push(t);
        };
        this.children().forEach(function (x) { return append(x); });
        return sorted.map(function (x) { return x.serializeToString(); }).join("\n");
    };
    TSAPIModule.prototype.modelClass = function () {
        return exports.MODEL_CLASS_API_MODULE;
    };
    return TSAPIModule;
}(TSModelElement));
exports.TSAPIModule = TSAPIModule;
var TSMember = (function (_super) {
    __extends(TSMember, _super);
    function TSMember() {
        _super.apply(this, arguments);
    }
    TSMember.prototype.modelClass = function () {
        return exports.MODEL_CLASS_MEMBER;
    };
    return TSMember;
}(TSModelElement));
exports.TSMember = TSMember;
var TSUnionTypeReference = (function (_super) {
    __extends(TSUnionTypeReference, _super);
    function TSUnionTypeReference() {
        _super.apply(this, arguments);
        this.array = function () { return false; };
    }
    TSUnionTypeReference.prototype.getFunctor = function () {
        return null;
    };
    //TODO FIXIT FIX IT WITH MIX IN
    TSUnionTypeReference.prototype.union = function (q) {
        var map = {};
        this.children().filter(function (x) { return x instanceof TSSimpleTypeReference; })
            .forEach(function (x) { return map[x.name] = true; });
        var gotNew = false;
        var flat = flattenUnionType(q);
        flat.forEach(function (x) {
            if (x instanceof TSSimpleTypeReference) {
                gotNew = gotNew || !map[q.name];
            }
            else {
                gotNew = true;
            }
        });
        if (!gotNew) {
            return this;
        }
        var r = new TSUnionTypeReference();
        this.children().forEach(function (x) { return r.addChild(x); });
        flat.forEach(function (x) { return r.addChild(x); });
        return r;
    };
    TSUnionTypeReference.prototype.isFunctor = function () {
        return false;
    };
    TSUnionTypeReference.prototype.canBeOmmited = function () {
        return false;
    };
    TSUnionTypeReference.prototype.serializeToString = function () {
        return this.children().map(function (x) { return x.serializeToString(); }).join(" | ");
    };
    TSUnionTypeReference.prototype.removeChild = function (child) { };
    TSUnionTypeReference.prototype.addChild = function (child) {
        this.children().push(child);
    };
    TSUnionTypeReference.prototype.modelClass = function () {
        return exports.MODEL_CLASS_UNION_TYPE_REFERENCE;
    };
    return TSUnionTypeReference;
}(TSModelElement));
exports.TSUnionTypeReference = TSUnionTypeReference;
var TSSimpleTypeReference = (function (_super) {
    __extends(TSSimpleTypeReference, _super);
    function TSSimpleTypeReference(p, tn) {
        var _this = this;
        _super.call(this, p);
        this.array = function () { return false; };
        this.genericStr = function () { return _this.typeParameters && _this.typeParameters.length > 0
            ? '<' + _this.typeParameters.map(function (p) { return p.serializeToString(); }).join(',') + '>'
            : ''; };
        this.name = tn;
    }
    TSSimpleTypeReference.prototype.isEmpty = function () {
        return false;
    };
    TSSimpleTypeReference.prototype.getFunctor = function () {
        return null;
    };
    TSSimpleTypeReference.prototype.canBeOmmited = function () {
        return false;
    };
    TSSimpleTypeReference.prototype.isFunctor = function () {
        return false;
    };
    TSSimpleTypeReference.prototype.union = function (q) {
        var _this = this;
        var flat = flattenUnionType(q);
        var gotThis = false;
        flat.filter(function (x) { return x instanceof TSSimpleTypeReference; }).map(function (x) { return x; })
            .forEach(function (x) { return gotThis = gotThis || (x.name == _this.name); });
        if (gotThis) {
            return q;
        }
        var r = new TSUnionTypeReference();
        r.addChild(this);
        flat.forEach(function (x) { return r.addChild(x); });
        return r;
    };
    TSSimpleTypeReference.prototype.serializeToString = function () {
        return this.name + this.genericStr();
    };
    TSSimpleTypeReference.prototype.modelClass = function () {
        return exports.MODEL_CLASS_SIMPLE_TYPE_REFERENCE;
    };
    return TSSimpleTypeReference;
}(TSModelElement));
exports.TSSimpleTypeReference = TSSimpleTypeReference;
var TSFunctionReference = (function (_super) {
    __extends(TSFunctionReference, _super);
    function TSFunctionReference(p) {
        var _this = this;
        _super.call(this, p);
        this.rangeType = new AnyType();
        this.parameters = [];
        this.array = function () { return false; };
        this.paramStr = function (appendDefault) {
            if (appendDefault === void 0) { appendDefault = false; }
            return '(' + _this.parameters
                .filter(function (x) { return !x.isEmpty(); })
                .map(function (p) { return p.serializeToString(appendDefault); })
                .join(', ') + ')';
        };
    }
    TSFunctionReference.prototype.isEmpty = function () {
        return false;
    };
    TSFunctionReference.prototype.getFunctor = function () {
        return null;
    };
    TSFunctionReference.prototype.canBeOmmited = function () {
        return false;
    };
    TSFunctionReference.prototype.isFunctor = function () {
        return true;
    };
    TSFunctionReference.prototype.union = function (q) {
        var r = new TSUnionTypeReference();
        r.addChild(this);
        r.addChild(q);
        return r;
    };
    TSFunctionReference.prototype.serializeToString = function () {
        return this.paramStr() + '=>' + this.rangeType.serializeToString();
    };
    TSFunctionReference.prototype.modelClass = function () {
        return exports.MODEL_CLASS_FUNCTION_REFERENCE;
    };
    return TSFunctionReference;
}(TSModelElement));
exports.TSFunctionReference = TSFunctionReference;
var TSArrayReference = (function (_super) {
    __extends(TSArrayReference, _super);
    function TSArrayReference(componentType) {
        if (componentType === void 0) { componentType = new AnyType(); }
        _super.call(this, exports.Universe);
        this.array = function () { return true; };
        this.componentType = componentType;
    }
    TSArrayReference.prototype.isEmpty = function () {
        return this.componentType ? true : false;
    };
    TSArrayReference.prototype.getFunctor = function () {
        return this.componentType.getFunctor();
    };
    TSArrayReference.prototype.canBeOmmited = function () {
        return false;
    };
    TSArrayReference.prototype.isFunctor = function () {
        return this.componentType.isFunctor();
    };
    TSArrayReference.prototype.union = function (q) {
        var r = new TSUnionTypeReference();
        r.addChild(this);
        r.addChild(q);
        return r;
    };
    TSArrayReference.prototype.serializeToString = function () {
        return this.componentType.serializeToString() + '[]';
    };
    TSArrayReference.prototype.modelClass = function () {
        return exports.MODEL_CLASS_ARRAY_REFERENCE;
    };
    return TSArrayReference;
}(TSModelElement));
exports.TSArrayReference = TSArrayReference;
var TSDeclaredInterfaceReference = (function (_super) {
    __extends(TSDeclaredInterfaceReference, _super);
    function TSDeclaredInterfaceReference(p, tn, _data) {
        _super.call(this, p, tn);
        this._data = _data;
    }
    TSDeclaredInterfaceReference.prototype.isEmpty = function () {
        return false;
    };
    TSDeclaredInterfaceReference.prototype.getFunctor = function () {
        return null;
    };
    TSDeclaredInterfaceReference.prototype.canBeOmmited = function () {
        return false;
    };
    TSDeclaredInterfaceReference.prototype.getOriginal = function () {
        return this._data;
    };
    TSDeclaredInterfaceReference.prototype.modelClass = function () {
        return exports.MODEL_CLASS_DECLARED_INTERFACE_REFERENCE;
    };
    return TSDeclaredInterfaceReference;
}(TSSimpleTypeReference));
exports.TSDeclaredInterfaceReference = TSDeclaredInterfaceReference;
var TSAnnotationReference = (function (_super) {
    __extends(TSAnnotationReference, _super);
    function TSAnnotationReference(p, tn, values) {
        if (values === void 0) { values = {}; }
        _super.call(this, p, tn);
        this.values = values;
    }
    TSAnnotationReference.prototype.value = function (key) {
        if (key === void 0) { key = 'value'; }
        return this.values[key];
    };
    return TSAnnotationReference;
}(TSSimpleTypeReference));
exports.TSAnnotationReference = TSAnnotationReference;
var TSDeclaredAnnotationReference = (function (_super) {
    __extends(TSDeclaredAnnotationReference, _super);
    function TSDeclaredAnnotationReference() {
        _super.apply(this, arguments);
        this.values = {};
    }
    TSDeclaredAnnotationReference.prototype.value = function (key) {
        if (key === void 0) { key = 'value'; }
        return this.values[key];
    };
    return TSDeclaredAnnotationReference;
}(TSDeclaredInterfaceReference));
exports.TSDeclaredAnnotationReference = TSDeclaredAnnotationReference;
var AnyType = (function (_super) {
    __extends(AnyType, _super);
    function AnyType(nm) {
        if (nm === void 0) { nm = "any"; }
        _super.call(this, exports.Universe, nm);
    }
    AnyType.prototype.union = function (q) {
        return q;
    };
    AnyType.prototype.modelClass = function () {
        return exports.MODEL_CLASS_ANY_TYPE_REFERENCE;
    };
    return AnyType;
}(TSSimpleTypeReference));
exports.AnyType = AnyType;
var TSStructuralTypeReference = (function (_super) {
    __extends(TSStructuralTypeReference, _super);
    function TSStructuralTypeReference(parent) {
        var _this = this;
        if (parent === void 0) { parent = exports.Universe; }
        _super.call(this, parent);
        this.array = function () { return false; };
        this.canBeOmmited = function () { return _this.locked ? false : _this.children().every(function (x) { return x.optional; }); };
    }
    TSStructuralTypeReference.prototype.visitReturnType = function (v) {
        //v.visitStructuralReturn(this);
        this.visit(v);
    };
    TSStructuralTypeReference.prototype.toReference = function () { return this; };
    TSStructuralTypeReference.prototype.union = function (q) {
        var r = new TSUnionTypeReference();
        r.addChild(this);
        r.addChild(q);
        return r;
    };
    TSStructuralTypeReference.prototype.serializeToString = function () {
        var body = this.children().map(function (x) { return ("\n" + x.serializeToString() + "\n"); }).join('');
        return "{" + body + "}";
    };
    TSStructuralTypeReference.prototype.modelClass = function () {
        return exports.MODEL_CLASS_STRUCTURAL_TYPE_REFERENCE;
    };
    return TSStructuralTypeReference;
}(TSTypeDeclaration));
exports.TSStructuralTypeReference = TSStructuralTypeReference;
(function (ParamLocation) {
    ParamLocation[ParamLocation["URI"] = 0] = "URI";
    ParamLocation[ParamLocation["BODY"] = 1] = "BODY";
    ParamLocation[ParamLocation["OPTIONS"] = 2] = "OPTIONS";
    ParamLocation[ParamLocation["OTHER"] = 3] = "OTHER";
})(exports.ParamLocation || (exports.ParamLocation = {}));
var ParamLocation = exports.ParamLocation;
var Param = (function (_super) {
    __extends(Param, _super);
    function Param(p, nm, location, tp, defaultValue) {
        if (tp === void 0) { tp = new TSSimpleTypeReference(exports.Universe, "string"); }
        _super.call(this, p);
        this.name = nm;
        this.ptype = tp;
        this.location = location;
        this.defaultValue = defaultValue;
    }
    Param.prototype.isEmpty = function () {
        return this.ptype.isEmpty();
    };
    Param.prototype.serializeToString = function (appendDefault) {
        if (appendDefault === void 0) { appendDefault = false; }
        //return this.name + (this.optional ? "?" : "") + ":" + this.ptype.serializeToString() + (this.ptype.canBeOmmited() ? "?" : "");
        return this.name + (this.optional || ((this.defaultValue != null) && !appendDefault) ? "?" : "")
            + (":" + this.ptype.serializeToString() + (this.ptype.canBeOmmited() ? "?" : ""))
            + (appendDefault && (this.defaultValue != null) ? '=' + JSON.stringify(this.defaultValue) : '');
    };
    Param.prototype.modelClass = function () {
        return exports.MODEL_CLASS_PARAM;
    };
    return Param;
}(TSModelElement));
exports.Param = Param;
var StringValue = (function (_super) {
    __extends(StringValue, _super);
    function StringValue(_value) {
        _super.call(this);
        this._value = _value;
    }
    StringValue.prototype.value = function () {
        return this._value;
    };
    StringValue.prototype.serializeToString = function () {
        return "\"" + this._value + "\"";
    };
    StringValue.prototype.modelClass = function () {
        return exports.MODEL_CLASS_STRING_VALUE;
    };
    return StringValue;
}(TSMember));
exports.StringValue = StringValue;
var ArrayValue = (function (_super) {
    __extends(ArrayValue, _super);
    function ArrayValue(_values) {
        _super.call(this);
        this._values = _values;
    }
    ArrayValue.prototype.value = function () {
        return this.serializeToString();
    };
    ArrayValue.prototype.serializeToString = function () {
        return "[ " + this._values.map(function (x) { return x.value(); }).join(', ') + " ]";
    };
    ArrayValue.prototype.values = function () {
        return this._values;
    };
    ArrayValue.prototype.modelClass = function () {
        return exports.MODEL_CLASS_ARRAY_VALUE;
    };
    return ArrayValue;
}(TSMember));
exports.ArrayValue = ArrayValue;
var TSAPIElementDeclaration = (function (_super) {
    __extends(TSAPIElementDeclaration, _super);
    function TSAPIElementDeclaration(p, name) {
        var _this = this;
        _super.call(this, p);
        this.rangeType = new AnyType();
        this.value = null;
        this.isStatic = false;
        this.paramStr = function (appendDefault) {
            if (appendDefault === void 0) { appendDefault = false; }
            return '( ' + _this.parameters
                .filter(function (x) { return !x.isEmpty(); })
                .map(function (p) { return _this.serializeParam(p, appendDefault); })
                .join(',') + ' )';
        };
        this.serializeParam = function (p, appendDefault) { return p.serializeToString(appendDefault); };
        this.isFunction = function () { return _this.parameters.length != 0 || _this.isFunc; };
        this.isAnonymousFunction = function () { return _this.isFunction() && _this.name === ''; };
        this.returnStr = function () { return _this.rangeType ? ':' + _this.rangeType.serializeToString() : ''; };
        this.name = name;
        this.parameters = [];
        this.rangeType = null;
        this.optional = false;
    }
    TSAPIElementDeclaration.prototype.visit = function (v) {
        if (v.startVisitElement(this)) {
            if (this.rangeType) {
                if (this.rangeType instanceof TSStructuralTypeReference && !this.isInterfaceMethodWithBody()) {
                    this.rangeType.visitReturnType(v);
                }
            }
            v.endVisitElement(this);
        }
    };
    TSAPIElementDeclaration.prototype.commentCode = function () {
        if (this.comment()) {
            return formattedComment(this._comment, 2);
        }
        return '';
    };
    TSAPIElementDeclaration.prototype.serializeToString = function (isImpl) {
        if (isImpl === void 0) { isImpl = false; }
        var x = (this.isPrivate ? 'private ' : '')
            + (this.isStatic ? 'static ' : '')
            + this.escapeDot(this.name)
            + (this.optional ? "?" : "")
            + (this.isFunction() ? this.paramStr(isImpl) : "")
            + this.returnStr();
        if (this.value) {
            x += '=' + this.value.value();
        }
        return this.commentCode() + x + (this.isFunction() && this.isInterfaceMethodWithBody() ? '' : this.body());
    };
    TSAPIElementDeclaration.prototype.body = function () {
        if (this._body == null)
            return "";
        return "{" + this._body + "}";
    };
    TSAPIElementDeclaration.prototype.escapeDot = function (name) {
        var escaped = exports.tsutil.escapeTypescriptPropertyName(name);
        return escaped;
    };
    TSAPIElementDeclaration.prototype.isInterfaceMethodWithBody = function () { return false; };
    TSAPIElementDeclaration.prototype.modelClass = function () {
        return exports.MODEL_CLASS_API_ELEMENT_DECLARATION;
    };
    return TSAPIElementDeclaration;
}(TSMember));
exports.TSAPIElementDeclaration = TSAPIElementDeclaration;
var TSConstructor = (function (_super) {
    __extends(TSConstructor, _super);
    function TSConstructor(p) {
        _super.call(this, p, 'constructor');
        this.serializeParam = function (p, appendDefault) { return 'protected ' + p.serializeToString(appendDefault); };
    }
    TSConstructor.prototype.modelClass = function () {
        return exports.MODEL_CLASS_CONSTRUCTOR;
    };
    return TSConstructor;
}(TSAPIElementDeclaration));
exports.TSConstructor = TSConstructor;
function flattenUnionType(ref) {
    var _this = this;
    if (!(ref instanceof TSUnionTypeReference)) {
        return [ref];
    }
    var map = {};
    var arr = [];
    ref.children().forEach(function (x) { return _this.flattenUnionType(x).forEach(function (y) {
        if (y instanceof TSSimpleTypeReference) {
            var st = y;
            var name = st.name;
            map[name] = st;
        }
        else {
            arr.push(y);
        }
    }); });
    return arr.concat(_.sortBy(Object.keys(map).map(function (x) { return map[x]; }), 'name'));
}
function formattedComment(str, indents) {
    var shift = '';
    for (var i = 0; i < indents; i++) {
        shift += '    ';
    }
    return "\n" + shift + "/**\n" + str.replace(/\*\//g, '* /').replace(/\r\n/g, '\n').split('\n').map(function (x) { return shift + ' * ' + x.trim(); }).join('\n') + "\n" + shift + " **/\n";
}
